var documenterSearchIndex = {"docs":
[{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"CurrentModule = BioFindr","category":"page"},{"location":"posteriorprobs/#Bayesian-inference-of-posterior-probabilities","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"","category":"section"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"After obtaining the PDFs for the LLRs from real data and the null hypotheses, we can convert LLR values into posterior probabilities P(mathcal H_mathrmaltmid mathrmLLR). We use a similar technique as in [Chen2007], which itself was based on a more general framework to estimate local FDRs in genome-wide studies [Storey2003]. This framework assumes that the real distribution of a certain test statistic forms a mixture distribution of null and alternative hypotheses. After estimating the null distribution, either analytically or by simulation, it can be compared against the real distribution to determine the proportion of null hypotheses, and consequently the posterior probability that the alternative hypothesis is true at any value of the statistic.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"[Chen2007]: Chen L, Emmert-Streib F, Storey J. Harnessing naturally randomized transcription to infer regulatory relationships among genes. Genome Biol 8, R219 (2007).","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"[Storey2003]: Storey JD, Tibshirani R. Statistical significance for genomewide studies. Proceedings of the National Academy of Sciences. 2003;100(16):9440–9445.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To be precise, consider an arbitrary likelihood ratio test. The fundamental assumption is that in the limit mathrmLLRrightarrow 0^+, all test cases come from the null hypothesis (mathcal H_mathrmnull), whilst as mathrmLLR increases, the proportion of alternative hypotheses (mathcal H_mathrmalt) also grows. The mixture distribution of real LLR values is assumed to have a PDF as","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"p(mathrmLLR) = P(mathcal H_mathrmnull) p(mathrmLLRmidmathcal H_mathrmnull) + P(mathcal H_mathrmalt)p(mathrmLLRmidmathcal H_mathrmalt)","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"The priors P(mathcal H_mathrmnull) and P(mathcal H_mathrmalt) sum to unity and correspond to the proportions of null and alternative hypotheses in the mixture distribution. For any test i=0dots5, Bayes' theorem then yields its posterior probability as","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\nP(mathcal H_mathrmalt^(i)midmathrmLLR^(i)) \n= fracp(mathrmLLR^(i)midmathcal H_mathrmalt^(i))p(mathrmLLR^(i))P(mathcal H_mathrmalt^(i))\n= 1 - fracp(mathrmLLR^(i)midmathcal H_mathrmnull^(i))p(mathrmLLR^(i))P(mathcal H_mathrmnull^(i))\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Based on this, we can define the posterior probabilities of the selected hypotheses according to the table in the Likelihood ratio tests section, i.e. the alternative for tests 0, 1, 2, 4, 5 and the null for test 3 as","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"P_iequiv begincases\n    P(mathcal H_mathrmalt^(i)midmathrmLLR^(i))  i=01245\n    P(mathcal H_mathrmnull^(i)midmathrmLLR^(i))i=3\nendcases","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Distributions can be estimated either separately for every (EA) pair or by pooling across all (EA) pairs. In practice, we test on the order of 10^3 to 10^4 candidate targets (\"B\") for every (EA) such that a separate conversion of LLR values to posterior probabilities is both feasible and recommended, as it accounts for different roles of every gene, especially hub genes, through different rates of alternative hypotheses.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Lastly, in a typical application of BioFindr, inputs of (EA) pairs will have been pre-determined as the set of significant eQTL-gene pairs from a genome-wide eQTL associaton analysis. In such cases, we may naturally assume P_1=1 for all considered pairs, and skip the primary test.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Posterior probabilities are computed in the pprob_col function:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"pprob_col","category":"page"},{"location":"posteriorprobs/#BioFindr.pprob_col","page":"Bayesian inference of posterior probabilities","title":"BioFindr.pprob_col","text":"pprob_col(Y::Matrix{T},Ycol::Vector{T}; method=\"moments\") where T<:AbstractFloat\n\nCompute the posterior probabilities for BioFindr test 0 (correlation test) for a given column vector Ycol against all columns of matrix Y.\n\nY and Ycol are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nThe optional parameter method determines the mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nSee also supernormalize, fit_mixdist_mom, fit_mixdist_KDE.\n\n\n\n\n\npprob_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}; method=\"moments\") where {T<:AbstractFloat, S<:Integer}\n\nCompute the posterior probabilities for the BioFindr causal tests for a given column vector Ycol with categorical instrument E against all columns of matrix Y: \n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nY is assumed to have undergone supernormalization with each column having mean zero and variance one.\n\nFor test 2, 4, and 5 the posterior probabilities are the probabilities of the alternative hypothesis being true. For test 3 they are the probabilities of the null hypothesis being true.\n\nThe optional parameter method determines the mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nSee also supernormalize, fit_mixdist_mom, fit_mixdist_KDE.\n\n\n\n\n\npprob_col(Y::Matrix{T},E::Vector{S}; method=\"moments\") where {T<:AbstractFloat, S<:Integer}\n\nCompute the posterior probabilities for differential expression of columns of matrix Y in the groups defined by  categorical vector E using BioFindr test 2 (Linkage test) \n\nY is assumed to have undergone supernormalization with each column having mean zero and variance one.\n\nThe optional parameter method determines the mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nSee also supernormalize, fit_mixdist_mom, fit_mixdist_KDE.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Computation of the posterior probabilities requires estimation of the prior probability P(mathcal H_mathrmnull) and fitting the mixture distribution p(mathrmLLR) from a set of random samples x_1 x_2dotsx_m.","category":"page"},{"location":"posteriorprobs/#Estimating-P({\\mathcal-H}_{\\mathrm{null}})","page":"Bayesian inference of posterior probabilities","title":"Estimating P(mathcal H_mathrmnull)","text":"","category":"section"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To estimate pi_0equiv P(mathcal H_mathrmnull) we convert the samples x_1 x_2dotsx_m to p-values under the null hypothesis,","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"p_i = P(Xgeq x_i mid mathcal H_mathrmnull) = int_x_i^infty p(x mid alpha_0beta_0) = mathrmccdf(x_imid alpha_0beta_0) ","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"where mathrmccdf denotes the complementary CDF. The histogram of p-values would show the characteristic shape of a set of anti-conservative p-values and pi_0 can be estimated by a robust bootstrap procedure:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"pi0est","category":"page"},{"location":"posteriorprobs/#BioFindr.pi0est","page":"Bayesian inference of posterior probabilities","title":"BioFindr.pi0est","text":"pi0est(pval)\n\nEstimate the proportion pi_0 of truly null features in a vector pval of p-values using a bootstrap method.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/#mom_postprobs","page":"Bayesian inference of posterior probabilities","title":"Method of moments estimation of the mixture distribution","text":"","category":"section"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To fit the mixture distribution, the original method[Wang2017] approximated PDFs with histograms. This requires proper choices of histogram bin widths, P(mathcal H_mathrmnull), and techniques to ensure the conversion from LLR to posterior probability is monotonically increasing and smooth.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"[Wang2017]: Wang L, Michoel T (2017) Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data. PLoS Comput Biol 13(8): e1005703.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"BioFindr uses an alternative, parametric approach, where the basic assumption is that the alternative distribution p(mathrmLLRmidmathcal H_mathrmalt) for each test also follows an LBeta distribution, that is, belongs to the same 2-parameter family of distributions as the null distributions. ","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Let us denote the random variable X=mathrmLLR taking values xgeq 0, and pi_0=P(mathcal H_mathrmnull). Our assumption is that","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"X sim pi_0 mathcalD(alpha_0beta_0) + (1-pi_0) mathcalD(alphabeta)","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"or","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"p(x) = pi_0 p(x mid alpha_0beta_0) + (1-pi_0) p(xmid alphabeta)","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"where mathcalD denotes an LBeta distribution with its PDF  p(xmid alphabeta) defined in Null distributions of the log-likelihood ratios. ","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Since the parameters alpha_0 and beta_0 of the null distribution are know exactly (see Null distributions of the log-likelihood ratios) and pi_0 can be estimated from the null p-values (see pi0est), it remains to estimate alpha, and beta from a set of random samples x_1 x_2dotsx_m of X.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To estimate alpha and beta we make use of the relation between the LBeta and Beta distributions (see Null distributions of the log-likelihood ratios). Define the random variable","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Y = 1 - e^-2 X","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Since by assumption each component of the mixture distribution of X follows an LBeta distribution, it follows that Y follows a mixture of Beta distributions,","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Y sim pi_0 BBigl(fracalpha_02fracbeta_02Bigr) + (1 - pi_0) BBigl(fracalpha2fracbeta2Bigr)","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"This result can be derived as follows. The function g(x)= 1-e^-2x increases monotonically for xgeq 0 and has the inverse g^-1(y)=-frac12ln(1-y). By the usual rules of transforming random variables:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\n    P(Yleq y) = P(1 - e^-2 X leq y) \n    = P(X leq g^-1(y))\n    = pi_0 P(X leq g^-1(y) mid alpha_0beta_0) + (1-pi_0) P(X leq g^-1(y) mid alphabeta)\n    = pi_0 F_0(g^-1(y)) + (1-pi_0) F_1(g^-1(y))\n    = pi_0 G_0(y) + (1-pi_0) G_1(y)\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"where F_0 and F_1 are the CDFs of random variables with distributions mathcalD(alpha_0beta_0) and mathcalD(alphabeta), respectively, and G_0 and G_1 are the CDFs of random variables with distributions B(alpha_02beta_02) and B(alpha2beta2). The last step used the transformation between LBeta and Beta distributions derived in  Null distributions of the log-likelihood ratios.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"The parameters of a Beta distribution relate to its first and second moment as follows: if Zsim B(ab), then","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\n    mathbbE(Z) = fracaa+b equiv m_1(ab)\n    mathbbE(Z^2) = mathbbE(Z) fraca+1a+b+1 equiv m_2(ab)\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Conversely, if first and second moments m_1 and m_2 are given, these equations can be inverted to give the parameters","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\n    a = fracm_1(m_1-m_2)(m_2-m_1^2)\n    b = frac(1-m_1)(m_1-m_2)(m_2-m_1^2)\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"which requires m_1m_2m_1^2","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Multiplying by two then gives the parameters of an LBeta distribution given the moments of the corresponding Beta distribution:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"fit_mom","category":"page"},{"location":"posteriorprobs/#BioFindr.fit_mom","page":"Bayesian inference of posterior probabilities","title":"BioFindr.fit_mom","text":"fit_mom(LBeta, m1, m2)\n\nFit an LBeta distribution to given first and second moments m1 and m2 of the corresponding Beta distribution. This requires that m1 and m2 satisfy the following relations (see also the Beta distribution wiki):\n\nm_1m_2 wedge m_2m_1^2\n\nAn AssertionError is thrown if the condition evaluates to false.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"For the mixture distributed Y, we have","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"mathbbE(Y^k) = pi_0 mathbbE(Y^k mid alpha_0beta_0) + (1-pi_0) mathbbE(Y^k mid alphabeta)","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"and hence","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\n    mathbbE(Y) = pi_0 m_1bigl(alpha_02beta_02bigr) + (1-pi_0) m_1bigl(alpha2beta2bigr)\n    mathbbE(Y^2) = pi_0 m_2bigl(alpha_02beta_02bigr) + (1-pi_0) m_2bigl(alpha2beta2bigr)\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"or","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"beginaligned\n    m_1bigl(alpha2beta2bigr) = fracmathbbE(Y) - pi_0 m_1bigl(alpha_02beta_02bigr)(1-pi_0)\n    m_2bigl(alpha2beta2bigr) = fracmathbbE(Y^2) - pi_0 m_2bigl(alpha_02beta_02bigr)(1-pi_0)\nendaligned","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Given samples x_1 x_2dotsx_m we transform them to","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"y_i = 1 - e^-2 x_i","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"and replace mathbbE(Y) and mathbbE(Y^2) in these equations by their estimates frac1nsum_i y_i and frac1nsum_i y_i^2, respectively. Likewise we replace pi_0 by its estimate hatpi_0 obtained as explained above (see pi0est). Since alpha_0 and beta_0 are known exactly, we obtain estimates hatm_1 and hatm_2 for the moments of a Beta distribution with parameters alpha2 and beta2. Plugging these moment estimates in the fit_mom function gives estimates hatalpha and hatbeta for the corresponding LBeta distribution.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To make this estimated distribution a valid alternative distribution component of the LLRs mixture distribution, we need to ensure the validity of two edge cases:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"In the limit mathrmLLRto 0^+, all cases must come from the null distribution. This will be the case if hatalpha geq alpha_0, and if this is not the case, we set hatalpha = alpha_0.\nIn the limit mathrmLLRto infty, all cases must come from the alternative distribution. This will be the case if hatbeta leq beta_0, and if this is not the case, we set hatbeta = beta_0.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"The entire procedure is implemented in the fit_mixdist_mom function:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"fit_mixdist_mom","category":"page"},{"location":"posteriorprobs/#BioFindr.fit_mixdist_mom","page":"Bayesian inference of posterior probabilities","title":"BioFindr.fit_mixdist_mom","text":"fit_mixdist_mom(llr,ns,ng=1,test=:corr)\n\nFit a two-component mixture distribution of two LBeta distributions to a vector of log-likelihood ratios llr using a method-of-moments algorithm. The first component is the true null distribution for a given BioFindr test with sample size ns and number of genotype groups ng. The second component is the alternative distribution, assumed to follow an LBeta distribution. The prior probability pi0 of an observation belonging to the null component is fixed and determined by the pi0est function. Hence only the parameters of the alternative component need to be estimated.\n\nThe input variable test can take the values:\n\n:corr - correlation test (test 0)\n:link - linkage test (test 1/2)\n:med - mediation test (test 3)\n:relev - relevance test (test 4)\n:pleio - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to :corr, the correlation test with ns samples is used and the third argument is ignored.\n\nSee also pi0est, fit_mom, LBeta, nulldist, nullpval.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/#kde_postprobs","page":"Bayesian inference of posterior probabilities","title":"Kernel density estimation of the mixture distribution","text":"","category":"section"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"The method of moments mixture distribution works very well in practice (by visual inspection of histograms and fitted distributions), but there is no theory to support it. Moreover the condition on the moments to produce valid distribution parameters (See fit_mom) is sometimes violated (typically when pi_0approx 1). Hence BioFindr also implements an alternative kernel density estimation method, comparable to the one implemented in the qvalue package for R.","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"fit_mixdist_KDE","category":"page"},{"location":"posteriorprobs/#BioFindr.fit_mixdist_KDE","page":"Bayesian inference of posterior probabilities","title":"BioFindr.fit_mixdist_KDE","text":"fit_mixdist_KDE(llr,ns,[ng,test])\n\nReturn posterior probabilities for a vector of log-likelihood ratio values llr for a given BioFindr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \":corr\", the correlation test with ns samples is used and the third argument is ignored.\n\nSee also pi0est, fit_kde, nulldist, nullpval.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"To mitigate edge effects due to LLRs being restricted to the half-open interval 0infty), the kernel density estimation of p(mathrmLLR) operates on transformed values:","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"fit_kde","category":"page"},{"location":"posteriorprobs/#BioFindr.fit_kde","page":"Bayesian inference of posterior probabilities","title":"BioFindr.fit_kde","text":"fit_kde(llr)\n\nFit a distribution function to a vector of log-likelihood ratio values llr using kernel density estimation. To avoid boundary effects in the KDE, the log-likelihoods (which take values in 0infty)) are first transformed to a vector\n\nz = log left( e^2 mathrmLLR - 1 right)\n\nwhich takes values in (-inftyinfty). KDE is applied to the transformed variables, and the probability density function (pdf) for llr is obtained from the pdf for z by the usual transformation rule for functions of random variables.\n\n\n\n\n\n","category":"function"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"Nevertheless, the KDE of p(mathrmLLR) will still fluctuate wildly near mathrmLLRapprox 0. Given a set of samples x_1dotsx_m with KDE estimated PDF values tildep(x_1) dots tildep(x_m), fit_mixdist_KDE calculates smoothened and monotonically increasing estimayes p(x_1)dots p(x_m) as","category":"page"},{"location":"posteriorprobs/","page":"Bayesian inference of posterior probabilities","title":"Bayesian inference of posterior probabilities","text":"p(x_i) = maxleft(0 min_jgeq i tildep(x_j) right)","category":"page"},{"location":"lbeta/","page":"-","title":"-","text":"CurrentModule = BioFindr","category":"page"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"CurrentModule = BioFindr","category":"page"},{"location":"utils/#Utilities","page":"Utilities","title":"Utilities","text":"","category":"section"},{"location":"utils/#Data-preparation","page":"Utilities","title":"Data preparation","text":"","category":"section"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"The findr function expects that input DataFrames use scientific types, mainly to distinguish between count-based expression data and categorical genotype data which could otherwise both be presented as integer-valued data. A utility function coerce_scitypes! is provided to convert data to the right scientific type:","category":"page"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"coerce_scitypes!\ntest_scitype","category":"page"},{"location":"utils/#BioFindr.coerce_scitypes!","page":"Utilities","title":"BioFindr.coerce_scitypes!","text":"coerce_scitypes!(df, scitype)\n\nCoerce all columns of dataframe df to the scientic type scitype. If df contains gene expression data, scitype can be  Continuous or Count. If df contains genotype data (or categorical data more generally), scitype can be Multiclass or OrderedFactor. Note though that genotypes are always treated as unordered categorical variables in BioFindr, and the ordering of levels for OrderedFactor data is not used. Continuous genotypes (e.g. expected allele counts outputted by genotype imputation methods) are not supported and must be converted to integers before calling this function.\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.test_scitype","page":"Utilities","title":"BioFindr.test_scitype","text":"test_scitype(df, scitype)\n\nTest if all columns of dataframe df have the scientific type scitype. If scitype is Continuous or Count, the function will return true if all columns are of that type. If scitype is Multiclass or OrderedFactor, the function will return true if all columns are of that type, although the columns may have different levels. If the columns have different types, the function will return false.\n\n\n\n\n\n","category":"function"},{"location":"utils/#Postprocessing-functions","page":"Utilities","title":"Postprocessing functions","text":"","category":"section"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"Several utility functions are used when findr is called with DataFrame inputs, some of which may be useful when manually post-processing the output of findr_matrix calls with matrix-based inputs.","category":"page"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"getpairs\nsymprobs\ncombineprobs\nstackprobs\nglobalfdr!\nglobalfdr\nqvalue","category":"page"},{"location":"utils/#BioFindr.getpairs","page":"Utilities","title":"BioFindr.getpairs","text":"getpairs(dX::T, dG::T, dE::T; colG=1, colX=2)\n\nGet pairs of indices of matching columns from dataframes dX and dG, with column names that should be matched listed in dataframe dE. The optional parameters colG (default value 1) and colX (default value 2) indicate which columns of dE need to be used for matching, either as a column number (integer) or column name (string). The optional parameter namesX can be used to match rows in dE to only a subset of the column names of dX.\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.symprobs","page":"Utilities","title":"BioFindr.symprobs","text":"symprobs(P; combination=\"prod\")\n\nSymmetrize a square matrix of posterior probabilities P. The optional parameter combination defines the symmetrization method:\n\nnone: do nothing (default)\nprod: P_ij=P_ijP_ji\nmean: P_ij=frac12(P_ij + P_ji)\nanti: P_ij=frac12(P_ij + 1 - P_ji)\n\nNote that the anti option defines \"antisymmetric\" probabilities, P_ij +  P_ji = 1, where evidence for a causal interaction ito j is also considered evidence against the opposite interaction jto i.\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.combineprobs","page":"Utilities","title":"BioFindr.combineprobs","text":"combineprobs(P; combination=\"none\")\n\nCombine posterior probabilities P for multiple likelihood likelihood ratio tests in a single probability (local precision) value.\n\nThe optional parameter combination defines the combination test:\n\nnone: do nothing, return the input P (default)\nmediation: the mediation test (P_2 P_3)\nIV: the instrumental variable or non-independence test (P_2 P_5)\norig: BioFindr's original combination (frac12(P_2 P_5 + P_4)\n\nThe input must be a three-dimensional array where the second dimension has size 4 and indexes the individual BioFindr tests (test 2-5). The output is a matrix of size size(P,1) x size(P,3).\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.stackprobs","page":"Utilities","title":"BioFindr.stackprobs","text":"stackprobs(P,colnames,rownames;nodiag=true)\n\nConvert a matrix of pairwise posterior probabilities P with column and row names colnames and rownames, respectively, to a stacked dataframe with Source, Target, and Probability columns, corresponding respectively to a column name, a row name, and the value of P in the corresponding row and column pair.\n\nThe optional parameter nodiag determines if self-interactions (equal row and column name) are excluded (nodiag=true, default) or not (nodiag=false).\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.globalfdr!","page":"Utilities","title":"BioFindr.globalfdr!","text":"globalfdr!(dP::T; FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nFor a DataFrame dP of posterior probabilities (local precision values), compute their corresponding q-values and keep only the rows with q-value less than a desired global false discovery rate FDR (default value 1, no selection). dP is assumed to be the output of a findr run with columns Source, Target, and Probability. The output DataFrame mirrors the structure of dP, keeping only the selected rows, and with an additional column qvalue. The output is sorted by qvalue if the optional argument sorted is true (default). If dP already contains a column qvalue, only the filtering and optional sorting are performed. \n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.globalfdr","page":"Utilities","title":"BioFindr.globalfdr","text":"globalfdr(P::Array{T},FDR) where T<:AbstractFloat\n\nFor an array (matrix or vector) P of posterior probabilities (local precision values), compute their corresponding q-values Q, and return the indices of P with q-value less than a desired global false discovery rate FDR.\n\nSee also qvalue\n\n\n\n\n\n","category":"function"},{"location":"utils/#BioFindr.qvalue","page":"Utilities","title":"BioFindr.qvalue","text":"qvalue(P::Vector{T}) where T<:AbstractFloat\n\nConvert a vector P of posterior probabilities (local precisions) to a vector of q-values. For a threshold value c on the posterior probabilities P, the global FDR, FDR(c) is defined as one minus the average local precision:\n\nFDR(c) = 1 - frac1N_c sum_icolon P_ileq c P_i\n\nwhere N_c=sharpicolon P_ileq c is the number of selected pairs. The q-value of a given index in P is then defined as the smallest FDR at which this pair is still called significant.\n\n\n\n\n\n","category":"function"},{"location":"utils/#Generating-simulated-data","page":"Utilities","title":"Generating simulated data","text":"","category":"section"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"BioFindr includes a function generate_test_data for generating simple simulated data for testing the package:","category":"page"},{"location":"utils/","page":"Utilities","title":"Utilities","text":"generate_test_data","category":"page"},{"location":"utils/#BioFindr.generate_test_data","page":"Utilities","title":"BioFindr.generate_test_data","text":"generate_test_data(nA, nB, fB, ns, ng, maf, bGA, bAB, supernormalize)\n\nGenerate test data for BioFindr with nA causal variables, nB potential target variables of which a random fraction fB are true targets for each causal variable, ns samples, ng genotype (instrumental variable) groups with minor allele frequence maf, and effect sizes bGA and bAB. Variables are sampled from a linear model with independent Gaussian noise with variance ϵ and correlated Gaussian noise with variance δ and covariance δρ. If supernormalize is true, the data is supernormalized.\n\n\n\n\n\n","category":"function"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"CurrentModule = BioFindr","category":"page"},{"location":"realLLR/#Likelihood-ratio-tests","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Consider correlated genes A, B, and a third variable E upstream of A and B, such as a significant eQTL of A. The eQTLs can be obtained either de novo using eQTL identification tools such as matrix-eQTL [Shabalin2012] or kruX [Qi2014], or from published analyses. Throughout this documentation, we assume that E is a significant eQTL of A, whereas extension to other data types is straightforward. We use A_i and B_i for the expression levels of gene A and B respectively, which are assumed to have gone through the supernormalization, and optionally the genotypes of the best eQTL of A as E_i, where i=1dotsn across samples. Genotypes are assumed to have a total of n_a categories, so E_iin0dotsn_a. We define the null and alternative hypotheses for a total of six tests, as shown in the table below. ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"[Shabalin2012]: Shabalin AA. Matrix eQTL: ultra fast eQTL analysis via large matrix operations, Bioinformatics, Volume 28, Issue 10, May 2012, Pages 1353–1358, ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"[Qi2014]: Qi J, Foroughi Asl H, Bjorkegren J, Michoel T. kruX: matrix-based non-parametric eQTL discovery. BMC Bioinformatics 15, 11 (2014).","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"note: Six likelihood ratio tests are performed to test the regulation A → B, numbered, named, and defined as shown.\n(Image: Six likelihood ratio tests)Copyright 2017 Wang, Michoel. DOI:10.1371/journal.pcbi.1005703.g001. Reused under the terms of the Creative Commons Attribution License","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"LLRs of every test are calculated separately as follows:","category":"page"},{"location":"realLLR/#corr_test_realLLR","page":"Likelihood ratio tests","title":"Correlation test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Define the null hypothesis as A and B are independent, and the alternative hypothesis as they are correlated:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmnull^mathrm(0)=Aqquad Bhspace4emmathcal H_mathrmalt^mathrm(0)=A - B","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The superscript (0) is the numbering of the test. Both hypotheses are modeled with gene expression levels following bivariate normal distributions, as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"beginpmatrix\n    A_i\n    B_i\nendpmatrix sim \nNleft(\n    beginpmatrix\n        0\n        0\n    endpmatrix\n    beginpmatrix\n        sigma_A0^2  rhosigma_A0sigma_B0\n        rhosigma_A0sigma_B0  sigma_B0^2\n    endpmatrix\nright)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"for i=1dotsn. The null hypothesis corresponds to rho=0.","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Maximum likelihood estimators (MLE) for the model parameters rho, sigma_A0, and sigma_B0 are","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"hatrho=frac1nsum_i=1^n A_i B_iqquadhatsigma_A0=hatsigma_B0=1","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"and the LLR is simply","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(0)=-fracn2log(1-hatrho^2)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"In the absence of genotype information, we use nonzero correlation between A and B as the indicator for Arightarrow B regulation, giving the posterior probability","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"P(A - B)=P(mathcalH_mathrmalt^mathrm(0) mid mathrmLLR^mathrm(0))","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The LLR for the correlation test of a specific gene A against all other genes B is implemented as a method of the realLLR_col function:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"realLLR_col(Y::Matrix{T},Ycol::Vector{T}) where T<:AbstractFloat","category":"page"},{"location":"realLLR/#BioFindr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T<:AbstractFloat","page":"Likelihood ratio tests","title":"BioFindr.realLLR_col","text":"realLLR_col(Y::Matrix{T},Ycol::Vector{T}) where T<:AbstractFloat\n\nCompute the log-likelihood ratios for BioFindr test 0 (correlation test) for a given column vector Ycol against all columns of matrix Y.\n\nY and Ycol are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize.\n\n\n\n\n\n","category":"method"},{"location":"realLLR/#prim_test_realLLR","page":"Likelihood ratio tests","title":"Primary linkage test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Verify that E regulates A from mathcal H_mathrmalt^mathrm(1)equiv Erightarrow A and mathcal H_mathrmnull^mathrm(1)equiv Eqquad A. For mathcal H_mathrmalt^mathrm(1), we model Erightarrow A as A follows a normal distribution whose mean is determined by E categorically, i.e.","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"A_imid E_isim N(mu_E_isigma_A^2)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"From the total likelihood p(Amid E)=prod_i=1^np(A_imid E_i), we find MLEs for model parameters mu_jj=01dotsn_a and sigma_A, as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"hatmu_j=frac1n_jsum_i=1^n A_i delta_E_i jquad hatsigma_A^2=1-sum_j=0^n_afracn_jn hatmu_j^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"where n_j is the sample count by genotype category,","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"n_j equiv sum_i=1^n delta_E_i j","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The Kronecker delta function is defined as delta_xy=1 for x=y, and 0 otherwise. When summing over all genotype values (j=0dotsn_a), we only pick those that exist (n_j0) throughout this article. Since the null hypothesis is simply that A_i is sampled from a genotype-independent normal distribution, with MLEs of mean zero and standard deviation one due to the supernormalization (see General inference algorithm), the LLR for test 1 becomes","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(1)=-fracn2lnhatsigma_A^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"By favoring a large mathrmLLR^mathrm(1), we select  mathcal H_mathrmalt^mathrm(1) and verify that E regulates A, with","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"P(Erightarrow A)=P(mathcal H_mathrmalt^mathrm(1)midmathrmLLR^mathrm(1))","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The LLR for the primary linkage test can be obtained by selecting the appropriate element from the results of the secondary linkage test.","category":"page"},{"location":"realLLR/#Secondary-linkage-test","page":"Likelihood ratio tests","title":"Secondary linkage test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The secondary test is identical with the primary test, except it verifies that E regulates B. Hence repeat the primary test on E and B and obtain the MLEs:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"hatnu_j=frac1n_jsum_i=1^n B_i delta_E_i jquad hatsigma_B^2=1-sum_j=0^n_afracn_jnhatnu_j^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"and the LLR as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(2)=-fracn2lnhatsigma_B^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmalt^mathrm(2) is chosen to verify that E regulates B.","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"In Association analysis, the linkage test is used standalone, and then its LLR for testing a specific grouping vector E against all genes B is implemented as a method of the realLLR_col function:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"realLLR_col(Y::Matrix{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}","category":"page"},{"location":"realLLR/#BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T<:AbstractFloat, S<:Integer}","page":"Likelihood ratio tests","title":"BioFindr.realLLR_col","text":"realLLR_col(Y::Array{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}\n\nCompute the log-likelihood ratios for BioFindr test 2 (Linkage test)  for a given categorical vector E against all columns of matrix Y.\n\nY is assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize, llrstats_col.\n\n\n\n\n\n","category":"method"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"In Causal inference, the LLR for the linkage test is computed together with the other tests for efficiency (see below).","category":"page"},{"location":"realLLR/#Conditional-independence-test","page":"Likelihood ratio tests","title":"Conditional independence test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Verify that E and B are independent when conditioning on A. This can be achieved by comparing ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmalt^mathrm(3)equiv Bleftarrow Erightarrow Awedge(Amathrm correlates with B)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"against ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmnull^mathrm(3)equiv Erightarrow Arightarrow B","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"LLRs close to zero then prefer mathcal H_mathrmnull^mathrm(3), and ensure that E regulates B only through A:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"P(Eperp Bmid A)=P(mathcal H_mathrmnull^mathrm(3)midmathrmLLR^mathrm(3))","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"For mathcal H_mathrmalt^mathrm(3), the bivariate normal distribution dependent on E can be represented as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"beginpmatrix\n    A_i\n    B_i\nendpmatrix\nmid E_i \nsim Nleft(\n    beginpmatrix\n        mu_E_i\n        nu_E_i\n    endpmatrix\n    beginpmatrix\n        sigma_A^2  rhosigma_Asigma_B \n        rhosigma_Asigma_B  sigma_B^2\n    endpmatrix\n    right)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"For mathcal H_mathrmnull^mathrm(3), the distributions follow A_imid E_i as in the Primary linkage test, as well as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"B_imid A_isim N(rho A_isigma_B^2)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Substituting parameters mu_jnu_jsigma_Asigma_Brho of mathcal H_mathrmalt^mathrm(3) and mu_jrhosigma_Asigma_B of mathcal H_mathrmnull^mathrm(3) with their MLEs, we obtain the LLR: ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(3) =-fracn2lnleft(hatsigma_A^2hatsigma_B^2-(hatrho+sigma_AB-1)^2right) + fracn2lnhatsigma_A^2 + fracn2ln(1-hatrho^2)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"where","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"sigma_AB equiv 1-sum_j=0^n_afracn_jnhatmu_jhatnu_j","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"and hatrho=frac1nsum_i=1^n A_iB_i is the same as in the Correlation test.","category":"page"},{"location":"realLLR/#Relevance-test","page":"Likelihood ratio tests","title":"Relevance test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Since the indirect regulation Erightarrow B tends to be weaker than any of its direct regulation components (Erightarrow A or Arightarrow B), we propose to test Erightarrow Arightarrow B with indirect regulation Erightarrow B as well as the direct regulation Arightarrow B for stronger distinguishing power on weak regulations. We define","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmalt^mathrm(4)equiv Erightarrow Awedge Erightarrow Bleftarrow A","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"and","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmnull^mathrm(4)equiv Erightarrow Aqquad B","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"This simply verifies that B is not independent from both A and E simultaneously. In the alternative hypothesis, B is regulated by E and A, which is modeled as a normal distribution whose mean is additively determined by E categorically and A linearly, i.e.","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"B_imid E_iA_isim N(nu_E_i+rho A_isigma_B^2)","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"We can hence solve its LLR as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(4)=-fracn2lnleft(hatsigma_A^2hatsigma_B^2-(hatrho+sigma_AB-1)^2right)+fracn2lnhatsigma_A^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"with all MLEs as defined before.","category":"page"},{"location":"realLLR/#Controlled-or-pleiotropy-test","page":"Likelihood ratio tests","title":"Controlled or pleiotropy test","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Based on the positives of the secondary test, we can further distinguish the alternative hypothesis ","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmalt^mathrm(5)equiv Bleftarrow Erightarrow Awedge Arightarrow B","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"from the null","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathcal H_mathrmnull^mathrm(5)equiv Bleftarrow Erightarrow A","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"to verify that E does not regulate A and B independently. Its LLR can be solved as","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"mathrmLLR^mathrm(5)=-fracn2lnleft(hatsigma_A^2hatsigma_B^2-(hatrho+sigma_AB-1)^2right)+fracn2lnhatsigma_A^2hatsigma_B^2","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"with all MLEs as defined before. Note that this test was called the controlled test in the original paper (see figure above), but we now prefer to refer to it as the pleiotropy test.","category":"page"},{"location":"realLLR/#Implementation","page":"Likelihood ratio tests","title":"Implementation","text":"","category":"section"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The LLRs for the secondary linkage, conditional independence, relevance, and pleiotrropy tests of a specific gene A against all other genes B can all be computed from the same sufficient statistics hatrho hatsigma_A^2 hatsigma_B^2 sigma_AB. Moreover note the relations","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"beginaligned\nmathrmLLR^mathrm(3) = mathrmLLR^mathrm(4) + fracn2ln (1-hatrho^2)\nmathrmLLR^mathrm(5) = mathrmLLR^mathrm(4) + fracn2ln hatsigma_B^2 = mathrmLLR^mathrm(4) - mathrmLLR^mathrm(2)\nendaligned","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"Hence for efficiency these LLRs are computed jointly by a method of the realLLR_col function:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"realLLR_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}","category":"page"},{"location":"realLLR/#BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T<:AbstractFloat, S<:Integer}","page":"Likelihood ratio tests","title":"BioFindr.realLLR_col","text":"realLLR_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}\n\nCompute the log-likelihood ratios for the BioFindr causal tests for a given column vector Ycol with categorical instrument E against all columns of matrix Y : \n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nY and Ycol are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize, llrstats_col.\n\n\n\n\n\n","category":"method"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"The MLEs of the various model parameters are computed in the llrstats_col and groupmeans functions:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"llrstats_col\ngroupmeans","category":"page"},{"location":"realLLR/#BioFindr.llrstats_col","page":"Likelihood ratio tests","title":"BioFindr.llrstats_col","text":"llrstats_col(Y,Ycol,E)\n\nCompute the sufficient statistics for the log-likelihood ratios for BioFindr tests 2-5  for a given column vector Ycol with categorical instrument E against all columns of Y.\n\nYcol and Y are assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nThe sufficient statistics are:\n\nthe covariance hatrho between Ycol and all columns of Y, returned in the first output argument,\nthe weighted average variances hatsigma_B^2 of each column of matrix Y over the groups (unique values) in E, returned as the first column of the second output argument,\nthe weighted average covariances sigma_AB between Ycol and all  columns of Y over the groups (unique values) of E, returned as the second column of the second output argument,\nthe weighted average variance hatsigma_A^2 of Ycol over the groups (unique values) in E, returned as the third output argument.\n\nSee also groupmeans.\n\n\n\n\n\nllrstats_col(Y,E)\n\nCompute the sufficient statistics for the log-likelihood ratios for BioFindr tests 2  for a given categorical vector E against all columns of Y.\n\nY is assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nThe sufficient statistics are the weighted average variances hatsigma_B^2 of each column of matrix Y over the groups (unique values) in E.\n\nSee also groupmeans.\n\n\n\n\n\n","category":"function"},{"location":"realLLR/#BioFindr.groupmeans","page":"Likelihood ratio tests","title":"BioFindr.groupmeans","text":"groupmeans(Y,Ycol,E)\n\nCompute the size n_j for each of the groups (unique values) in categorical vector E (first output argument), and the means hatnu_j of each column of matrix Y (second output argument) and hatmu_j of Ycol (third output argument) for each of the groups in E.\n\n\n\n\n\ngroupmeans(Y,E)\n\nCompute the size n_j for each of the groups (unique values) in categorical vector E (first output argument), and the means hatnu_j of each column of matrix Y (second output argument) for each of the groups in E.\n\n\n\n\n\n","category":"function"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"A summary of all methods of the realLLR_col function:","category":"page"},{"location":"realLLR/","page":"Likelihood ratio tests","title":"Likelihood ratio tests","text":"realLLR_col","category":"page"},{"location":"realLLR/#BioFindr.realLLR_col","page":"Likelihood ratio tests","title":"BioFindr.realLLR_col","text":"realLLR_col(Y::Matrix{T},Ycol::Vector{T}) where T<:AbstractFloat\n\nCompute the log-likelihood ratios for BioFindr test 0 (correlation test) for a given column vector Ycol against all columns of matrix Y.\n\nY and Ycol are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize.\n\n\n\n\n\nrealLLR_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}\n\nCompute the log-likelihood ratios for the BioFindr causal tests for a given column vector Ycol with categorical instrument E against all columns of matrix Y : \n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nY and Ycol are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize, llrstats_col.\n\n\n\n\n\nrealLLR_col(Y::Array{T},E::Vector{S}) where {T<:AbstractFloat, S<:Integer}\n\nCompute the log-likelihood ratios for BioFindr test 2 (Linkage test)  for a given categorical vector E against all columns of matrix Y.\n\nY is assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).\n\nSee also supernormalize, llrstats_col.\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"CurrentModule = BioFindr","category":"page"},{"location":"randomLLR/#Null-distributions-of-the-log-likelihood-ratios","page":"Null distributions of the LLRs","title":"Null distributions of the log-likelihood ratios","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The null distribution of an LLR, p(mathrmLLRmidmathcal H_mathrmnull), may be obtained either by simulation or analytically. Simulation, such as random permutations from real data or the generation of random data from statistics of real data, can deal with a much broader range of scenarios in which analytical expressions are unattainable. However, the drawbacks are obvious: simulation can take hundreds of times longer than analytical methods to reach a satisfiable precision. Here we obtained analytical expressions of p(mathrmLLRmidmathcal H_mathrmnull) for all the Likelihood ratio tests.","category":"page"},{"location":"randomLLR/#Correlation-test","page":"Null distributions of the LLRs","title":"Correlation test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathcal H_mathrmnull^mathrm(0)=Aquad B indicates no correlation between A and B. Therefore, we can start from","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"tildeB_isimmathrmiid N(01)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"In order to simulate the supernormalization step, we normalize tildeB_i into B_i with zero mean and unit variance as:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"B_iequivfractildeB_i-bartildeB_isigma_tildeBhspace3embartildeBequiv frac1nsum_i=1^ntildeB_ihspace3emsigma_tildeB^2equivfrac1nsum_i=1^nleft(tildeB_i-bartildeBright)^2","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Transform the random variables tildeB_i by defining  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"beginaligned\n    X_1 equiv frac1sqrtnsum_i=1^nA_itildeB_i\n    X_2 equiv frac1sqrtnsum_i=1^ntildeB_i\n    X_3 equiv left(sum_i=1^ntildeB_i^2right)-X_1^2-X_2^2\nendaligned","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Since tildeB_isimmathrmiid N(01), we can easily verify that X_1X_2X_3 are independent, and  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"X_1sim N(01)hspace3emX_2sim N(01)hspace3emX_3simchi^2(n-2)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Expressing the LLR for the correlation test in terms of X_1X_2X_3 gives  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(0)=-fracn2ln(1-Y)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"in which","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"YequivfracX_1^2X_1^2+X_3simmathrmBetaBigl(frac 12fracn-22Bigr)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"follows the Beta distribution.","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"We define the distribution mathcal D(alphabeta) as the distribution of a random variable Z=-frac 12ln(1-Y) for YsimmathrmBeta(alpha2beta2), i.e.  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Z=-frac12ln(1-Y)simmathcal D(alphabeta)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The probability density function (PDF) for Zsimmathcal D(alphabeta) can be derived as: for z0, ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"p(zmid alphabeta)=frac2B(alpha2beta2)left(1-e^-2zright)^(alpha2-1)e^-beta z","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"and for zle0, p(zmid alphabeta)=0. Here B(ab) is the Beta function. Therefore the null distribution for the correlation test is simply ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(0)nsimmathcal D(1n-2)","category":"page"},{"location":"randomLLR/#Primary-linkage-test","page":"Null distributions of the LLRs","title":"Primary linkage test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathcal H_mathrmnull^mathrm(1)=Ehspace16emA indicates no regulation from E to A. Therefore, similarly with the correlation test, we start from  tildeA_isimmathrmiid N(01) and normalize them to A_i with zero mean and unit variance.","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The expression of mathrmLLR^mathrm(1) then becomes:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(1) = -fracn2lnleft(1-sum_j=0^n_afracn_jnfracleft(hattildemu_j-bartildeAright)^2sigma_tildeA^2right)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"where","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"hattildemu_jequivfrac1n_jsum_i=1^ntildeA_idelta_E_ij","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"For now, assume all possible genotypes are present, i.e. n_j0 for j=0dotsn_a. Transform tildeA_i by defining  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"beginaligned\n    X_j equiv sqrtn_jhattildemu_jhspace9emmathrmfor j=0dotsn_a\n    X_n_a+1 equiv Biggl(sum_i=1^ntildeA_i^2Biggr)-Biggl(sum_j=0^n_aX_j^2Biggr)\nendaligned","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Then we can similarly verify that X_i are pairwise independent, and ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"beginaligned\n    X_i sim N(01) mathrmfor i=0dotsn_a\n    X_n_a + 1 sim chi^2(n-n_a-1)\nendaligned","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Again transform X_i by defining independent random variables","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"beginaligned\n    Y_1 equiv sum_j=0^n_asqrtfracn_jnX_jsim N(01)\n    Y_2 equiv left(sum_j=0^n_aX_j^2right)-Y_1^2simchi^2(n_a)\n    Y_3 equiv X_n_a+1simchi^2(n-n_a-1)\nendaligned","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Some calculation would reveal","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(1)=-fracn2lnleft(1-fracY_2Y_2+Y_3right)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"i.e.","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(1)nsimmathcal D(n_an-n_a-1)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"To account for genotypes that do not show up in the samples, define n_vequivsum_jinjmid n_j01 as the number of different genotype values across all samples. Then  ","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(1)nsimmathcal D(n_v-1n-n_v)","category":"page"},{"location":"randomLLR/#Secondary-linkage-test","page":"Null distributions of the LLRs","title":"Secondary linkage test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Since the null hypotheses and LLRs of primary and secondary tests are identical, mathrmLLR^mathrm(2) follows the same null distribution as mathrmLLR^mathrm(1).","category":"page"},{"location":"randomLLR/#Conditional-independence-test","page":"Null distributions of the LLRs","title":"Conditional independence test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The conditional independence test verifies if E and B are uncorrelated when conditioning on A, with mathcal H_mathrmnull^mathrm(3)=Erightarrow Arightarrow B. For this purpose, we keep E and A intact while randomizing tildeB_i according to B's correlation with A:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"tildeB_iequivhatrho A_i+sqrt1-hatrho^2X_ihspace2emX_isimmathrmiid N(01)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Then tildeB_i is normalized to B_i as before. The null distribution of mathrmLLR^mathrm(3) can be obtained with similar but more complex computations as before, as","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(3)nsimmathcal D(n_v-1n-n_v-1)","category":"page"},{"location":"randomLLR/#Relevance-test","page":"Null distributions of the LLRs","title":"Relevance test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The null distribution of mathrmLLR^mathrm(4) can be obtained similarly by randomizing B_i, as","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(4)nsimmathcal D(n_vn-n_v-1)","category":"page"},{"location":"randomLLR/#Pleiotropy-test","page":"Null distributions of the LLRs","title":"Pleiotropy test","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"To compute the null distribution for the controlled test, we start from","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"tildeB_i=hatnu_E_i+hatsigma_B X_ihspace2emX_isim N(01)","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"and then normalize tildeB_i into B_i. Some calculation reveals the null distribution as","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"mathrmLLR^mathrm(5)nsimmathcal D(1n-n_v-1)","category":"page"},{"location":"randomLLR/#Implementation","page":"Null distributions of the LLRs","title":"Implementation","text":"","category":"section"},{"location":"randomLLR/#Null-distribution-family","page":"Null distributions of the LLRs","title":"Null distribution family","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The family of distributions mathcalD, termed LBeta in the code, is implemented as a continuous, univariate distribution type using the Distributions package:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"LBeta","category":"page"},{"location":"randomLLR/#BioFindr.LBeta","page":"Null distributions of the LLRs","title":"BioFindr.LBeta","text":"LBeta(α,β)\n\nThe LBeta distribution with parameters alpha and beta is defined as the distribution of a random variable \n\nX=-frac12(ln(1-Y))\n\nwhere YsimoperatornameBeta(alpha2 beta2)\n\n\n\n\n\n","category":"type"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Its PDF is implemented explicitly from the definition above:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"pdf\nlogpdf","category":"page"},{"location":"randomLLR/#Distributions.pdf","page":"Null distributions of the LLRs","title":"Distributions.pdf","text":"pdf(d,x)\n\nEvaluate the probability density function of an LBeta distribution with support on xgeq 0. \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#Distributions.logpdf","page":"Null distributions of the LLRs","title":"Distributions.logpdf","text":"logpdf(d,x)\n\nEvaluate the probability density function of an LBeta distribution.  \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"The CDF is implemented by calling the corresponding functions for a Beta distribution:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"cdf\nccdf\nlogccdf","category":"page"},{"location":"randomLLR/#Distributions.cdf","page":"Null distributions of the LLRs","title":"Distributions.cdf","text":"cdf(d,x)\n\nEvaluate the cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#Distributions.ccdf","page":"Null distributions of the LLRs","title":"Distributions.ccdf","text":"ccdf(d,x)\n\nEvaluate the complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#Distributions.logccdf","page":"Null distributions of the LLRs","title":"Distributions.logccdf","text":"logccdf(d,x)\n\nEvaluate the log complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#Test-specific-null-distributions","page":"Null distributions of the LLRs","title":"Test-specific null distributions","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"To construct a null distribution object for a specific test, use:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"nulldist","category":"page"},{"location":"randomLLR/#BioFindr.nulldist","page":"Null distributions of the LLRs","title":"BioFindr.nulldist","text":"nulldist(ns,[ng,test])\n\nReturn an LBeta distribution that is the null distribution of the log-likelihood ratio for a given BioFindr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n:corr - correlation test (test 0)\n:link - linkage test (test 1/2)\n:med - mediation test (test 3)\n:relev - relevance test (test 4)\n:pleio - pleiotropy test (test 5)\n\nWith only one input argument, the null distribution for the correlation test with ns samples is returned. With two input arguments, or  with three arguments and test equal to :corr, the null distribution for the correlation test with ns samples is returned and the second argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Convenience functions are provided that wrap around the pdf, ccdf, or logccdf functions:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"nullpdf\nnullpval\nnulllog10pval","category":"page"},{"location":"randomLLR/#BioFindr.nullpdf","page":"Null distributions of the LLRs","title":"BioFindr.nullpdf","text":"nullpdf(llr,ns,[ng,test])\n\nReturn probability distribution function evaluations for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given BioFindr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n:corr - correlation test (test 0)\n:link - linkage test (test 1/2)\n:med - mediation test (test 3)\n:relev - relevance test (test 4)\n:pleio - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to :corr, the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#BioFindr.nullpval","page":"Null distributions of the LLRs","title":"BioFindr.nullpval","text":"nullpval(llr,ns,[ng,test])\n\nReturn p-values for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given BioFindr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n:corr - correlation test (test 0)\n:link - linkage test (test 1/2)\n:med - mediation test (test 3)\n:relev - relevance test (test 4)\n:pleio - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to :corr, the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#BioFindr.nulllog10pval","page":"Null distributions of the LLRs","title":"BioFindr.nulllog10pval","text":"nulllog10pval(llr,ns,[ng,test])\n\nReturn negative log10 p-values for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given BioFindr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n:corr - correlation test (test 0)\n:link - linkage test (test 1/2)\n:med - mediation test (test 3)\n:relev - relevance test (test 4)\n:pleio - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to :corr, the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#LBeta-related-functions","page":"Null distributions of the LLRs","title":"LBeta related functions","text":"","category":"section"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"Some helper functions to work with LBeta distribution objects:","category":"page"},{"location":"randomLLR/","page":"Null distributions of the LLRs","title":"Null distributions of the LLRs","text":"compute_moments\ntest_moments\nfit\nfit_mle","category":"page"},{"location":"randomLLR/#BioFindr.compute_moments","page":"Null distributions of the LLRs","title":"BioFindr.compute_moments","text":"compute_moments(d::LBeta)\n\nCompute the first and second moments m1 and m2 of the Beta distribution corresponding to the given LBeta distribution. \n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#BioFindr.test_moments","page":"Null distributions of the LLRs","title":"BioFindr.test_moments","text":"test_moments(m1,m2)\n\nTest if the given first and second moments m1 and m2 satisfy the conditions for a valid Beta distribution. (see also the Beta distribution wiki):\n\n``m_1>m_2 \\wedge m_2>m_1^2``\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#BioFindr.fit","page":"Null distributions of the LLRs","title":"BioFindr.fit","text":"fit(LBeta, x)\n\nFit an LBeta distribution to data x using the method of moments by exploiting its relationship to the Beta distribution.\n\nSee https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl\n\n\n\n\n\n","category":"function"},{"location":"randomLLR/#Distributions.fit_mle","page":"Null distributions of the LLRs","title":"Distributions.fit_mle","text":"fit_mle(LBeta, x)\n\nFit an LBeta distribution to data x using maximum-likelihood estimation by exploiting its relationship to the Beta distribution.\n\nSee https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl\n\n\n\n\n\n","category":"function"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"CurrentModule = BioFindr","category":"page"},{"location":"testLLR/#Tests-to-evaluate","page":"Tests to evaluate","title":"Tests to evaluate","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"Based on the six Likelihood ratio tests, we use the following tests and test combinations for the inference of genetic regulations:","category":"page"},{"location":"testLLR/#Coexpression-analysis","page":"Tests to evaluate","title":"Coexpression analysis","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The correlation test is introduced as a benchmark, against which we can compare other methods involving genotype information. Pairwise correlation is a simple measure for the probability of two genes being functionally related either through direct or indirect regulation, or through coregulation by a third factor. Bayesian inference additionally considers different gene roles. Its predicted posterior probability for regulation is P_0.","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"Correlation analysis can be performed by calling findr with one argument, a dataframe of gene expression values:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr(dX::T) where T<:AbstractDataFrame","category":"page"},{"location":"testLLR/#BioFindr.findr-Tuple{T} where T<:DataFrames.AbstractDataFrame","page":"Tests to evaluate","title":"BioFindr.findr","text":"findr(dX::T; colnames=[], method=\"moments\", FDR=1.0, sorted=true, combination=\"none\") where T<:AbstractDataFrame\n\nWrapper for findr_matrix(Matrix(dX)) when the input dX is in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target, (Posterior) Probability, and qvalue columns. All columns of the input DataFrame must have a count or continuous scientific type. \n\nThe optional parameter colnames (vector of strings) determines whether we consider all columns of dX as source nodes (colnames=[], default), or only a subset of columns determined by the variable names in the vector colnames.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nThe optional parameter combination determines whether the output must be symmetrized. Possible values are none (default), prod, mean, or anti. If the optional parameter colnames is non-empty, symmetrization makes no sense and an error will be thrown unless combination=\"none\".\n\nSee also coerce_scitypes!, findr_matrix(::Matrix), symprobs, stackprobs, globalfdr!.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"If you prefer to work with matrix-based data, the equivalent function findr_matrix:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr_matrix(X::Matrix{T}) where T<:AbstractFloat","category":"page"},{"location":"testLLR/#BioFindr.findr_matrix-Union{Tuple{Matrix{T}}, Tuple{T}} where T<:AbstractFloat","page":"Tests to evaluate","title":"BioFindr.findr_matrix","text":"findr_matrix(X::Matrix{T}; cols=[], method=\"moments\", combination=\"none\") where T<:AbstractFloat\n\nCompute posterior probabilities for nonzero pairwise correlations between columns of input matrix X. The probabilities are directed (asymmetric) in the sense that they are estimated from a column-specific background distribution.\n\nThe optional parameter cols (vector of integers) determines whether we consider all columns of X as source nodes (cols=[], default), or only a subset of columns determined by the indices in the vector cols.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter combination determines whether the output must be symmetrized. Possible values are none (default), prod, mean, or anti. If the optional parameter cols is non-empty, symmetrization makes no sense and an error will be thrown unless combination=\"none\".\n\nSee also findr(::DataFrame), symprobs, supernormalize, pprob_col.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/#Association-analysis","page":"Tests to evaluate","title":"Association analysis","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The secondary linkage test is introduced to test association between genetic variants and gene expression levels, and can be used more generally to analyze differential expression of genes across groups defined by any kind of categorical variable. Its predicted posterior probability for differential expression is P_2.","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"Association analysis can be performed by calling findr with two arguments, dataframes of continuous gene expression values and categorical genotype or more general grouping values, respectively:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr(dX::T, dG::T) where T<:AbstractDataFrame","category":"page"},{"location":"testLLR/#BioFindr.findr-Union{Tuple{T}, Tuple{T, T}} where T<:DataFrames.AbstractDataFrame","page":"Tests to evaluate","title":"BioFindr.findr","text":"findr(dX::T, dG::T; method=\"moments\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr_matrix(Matrix(dX), Matrix(dG)) when the inputs dX and dG are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target (Posterior) Probability, and qvalue columns.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nNote that depending on the scitype of dG, different matrix-based methods are called. If the scitype dG is Count or Continuous, posterior probabilities for nonzero pairwise correlations between the variables in dG and variables in dX are computed. If the scitype of dG is Multiclass or OrderedFactor, posterior probabilities for nonzero differential expression of variables in dX across groups defined by the variables in dG are computed\n\nSee also coerce_scitypes!, findr_matrix(::Matrix,::Array), stackprobs, globalfdr!. \n\n\n\n\n\n","category":"method"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"or with matrix-based inputs:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr_matrix(X::Matrix{T}, G::Array{S}) where {T<:AbstractFloat, S<:Integer}","category":"page"},{"location":"testLLR/#BioFindr.findr_matrix-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{S}}} where {T<:AbstractFloat, S<:Integer}","page":"Tests to evaluate","title":"BioFindr.findr_matrix","text":"findr_matrix(X::Matrix{T},G::Array{S}; method=\"moments\") where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero differential expression of colunns of input matrix X across groups defined by one or more categorical variables (columns of G).\n\nReturn a matrix of size ncols(X) x ncols(G)\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nSee also findr(::DataFrame,::DataFrame), supernormalize, pprob_col.\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. CategoricalArrays will be supported in the future.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/#Causal-inference","page":"Tests to evaluate","title":"Causal inference","text":"","category":"section"},{"location":"testLLR/#Mediation","page":"Tests to evaluate","title":"Mediation","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The traditional causal inference test, as explained in [Chen2007], suggested that the regulatory relation Eto Ato Bcan be confirmed with the combination of three separate tests: E regulates A, E regulates B, and E only regulates B through A (i.e. E and B become independent when conditioning on A). They correspond to the primary, secondary, and conditional independence tests respectively. The regulatory relation Eto Ato B is regarded positive only when all three tests return positive. The three tests filter the initial hypothesis space of all possible relations between E, A, and B, sequentially to Eto A (primary test), Eto A wedge Eto B (secondary test), and Eto Ato B wedge (no confounder for A and B) (conditional independence test). The resulting test is stronger than Eto Ato B by disallowing confounders for A and B. So its probability can be broken down as","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"P_textmed equiv P_1P_2P_3","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"BioFindr expects a set of significant eQTLs and their associated genes as input, and therefore P_1=1 is assured and not calculated separately in BioFindr. Note that P_textmed is the estimated local precision, i.e. the probability that tests 2 and 3 are both true. Correspondinly, its local FDR (the probability that one of them is false) is 1-P_textmed.","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"[Chen2007]: Chen L, Emmert-Streib F, Storey J. Harnessing naturally randomized transcription to infer regulatory relationships among genes. Genome Biol 8, R219 (2007).","category":"page"},{"location":"testLLR/#Instrumental-variables","page":"Tests to evaluate","title":"Instrumental variables","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The pleiotropy test is introduced to test if an Eto B association is not independent of the Eto A association, that is, if an independent pleiotropic effect of E on both genes can be excluded. If E regulates A (is a cis-eQTL for A), and E regulates B, and  B and A are not independent given E, then we can regard E as a proxy or instrumental variable for A and infer a regulatory relation Eto Ato B from the Eto B association. The three tests verify the hypothesis that B leftarrow E to A wedge lnot(A  B  E), a superset of Eto Ato B.  Its probability can be broken down as","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"P_textIV equiv P_1P_2P_5","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"As before, P_1=1 is assured and not calculated separately in BioFindr. P_textIV is again the estimated local precision, i.e. the probability that tests 2 and 5 are both true, and its local FDR (the probability that one of them is false) is 1-P_textIV.","category":"page"},{"location":"testLLR/#Relevance","page":"Tests to evaluate","title":"Relevance","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The relevance test is introduced to address weak interactions that are undetectable by the secondary test from existing data (P_2 close to 0). This term still grants higher-than-null significance to weak interactions, and verifies that Eto A wedge (Eto B vee A - B), also a superset of Eto Ato B. Its probability can be broken down as","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"P_textrelev equiv P_1P_4","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The original Findr paper proposed to combine the instrumental variable and relevance tests in a novel test whose probability can be broken down as","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"P_textorig equiv frac12 P_1 bigl( P_4 + P_2P_5) = frac12bigl( P_textrelev + P_textIV bigr)","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"In the extreme undetectable limit where P_2 = 0 but P_4 neq 0, the novel test automatically reduces to one half of the relevance test, which assumes equal probability of either direction and assigns half of the relevance test probability to A to B.","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"The composite design of the novel test aims not to miss any genuine regulation whilst distinguishing the full spectrum of possible interactions. When the signal level is too weak for tests 2 and 5, we expect P_4 to still provide distinguishing power better than random predictions. When the interaction is strong, P_2 P_5 is then able to pick up true targets regardless of the existence of hidden confounders.","category":"page"},{"location":"testLLR/#Implementation","page":"Tests to evaluate","title":"Implementation","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"Causal inference can be performed by calling findr with three arguments, dataframes of gene expression and genotype values, and a mapping of matching (EA) pairs; the preferred test can be set through the combination parameter:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr(dX::T, dG::T, dE::T; colG=1, colX=2, combination=\"IV\") where T<:AbstractDataFrame","category":"page"},{"location":"testLLR/#BioFindr.findr-Union{Tuple{T}, Tuple{T, T, T}} where T<:DataFrames.AbstractDataFrame","page":"Tests to evaluate","title":"BioFindr.findr","text":"findr(dX::T, dG::T, dE::T; colX=2, colG=1, method=\"moments\", combination=\"IV\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr(Matrix(dX), Matrix(dG), pairGX) when the inputs are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target (Posterior) Probability, and qvalue columns. When DataFrames are used, only combined posterior probabilities can be returned (combination=\"IV\" (default), \"mediation\", or \"orig\").\n\nThe input dataframes are:\n\ndX - DataFrame with expression data, columns are genes\ndG - DataFrame with genotype data, columns are variants (SNPs)\ndE - DataFrame with eQTL results, must contain columns with gene and SNP IDs that can be mapped to column names in dX and dG, respectively\n\nThe numeric mapping between column indices in Matrix(dG) and Matrix(dX) is obtained from these inputs using the getpairs function and the optional parameters:\n\ncolG - name or number of variant ID column in dE, default 1\ncolX - name or number of gene ID column in dE, default 2\nnamesX - names of a possible subset of columns in dX to be considered as potential causal regulators (default names(dX))\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nSee also findr(::Matrix,::Array,::Matrix), getpairs, combineprobs, stackprobs, globalfdr!.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"or with matrix-based input data:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr_matrix(X::Matrix{T},G::Matrix{S},pairGX::Matrix{R}; combination=\"none\") where {T<:AbstractFloat, S<:Integer, R<:Integer}","category":"page"},{"location":"testLLR/#BioFindr.findr_matrix-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{R}}} where {T<:AbstractFloat, S<:Integer, R<:Integer}","page":"Tests to evaluate","title":"BioFindr.findr_matrix","text":"findr_matrix(X::Matrix{T},G::Matrix{S},pairGX::Matrix{S}; method=\"moments\", combination=\"none\") where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero causal relations between columns of input matrix X. The probabilities are estimated for relations going from a subset of columns of X that have a (discrete) instrumental variable in input matrix G to all columns of X, while excluding self-interactions (given default value 1). The matching between columns of X and columns of G is given by pairGX, a two-column array where the first column corresponds to a column index in G and the second to a column index in X.\n\nPosterior probabilities are computed for the following tests\n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nwhich can be combined into the mediation test (P_2 P_3; combination=\"mediation\"), the instrumental variable or non-independence test (P_2 P_5; combination=\"IV\"), or BioFindr's original combination (frac12(P_2 P_5 + P_4); combination=\"orig\"). By default, individual probability matrices for all tests are returned (combination=\"none\").\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nIf combination=\"none\", then the output has size ncols(X) x 4 x ncols(G), where the middle index indexes the tests, and otherwise the output has size ncols(X) x ncols(G). \n\nSee also findr(::DataFrame,::DataFrame,::DataFrame), supernormalize, pprob_col, combineprobs.\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/#Bipartite-causal-inference","page":"Tests to evaluate","title":"Bipartite causal inference","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"In the general case, we assume that there is one set of genes, of which the set of A-genes (genes with matching instrument E) is a subset, and that all possible directed regulations are tested. In some situations we are instead searching for a bipartite network from one set of potential causal factors (e.g. micro-RNAs) to another set of potential targets (e.g. protein-coding genes). In this case, causal inference can be performed by calling findr with four arguments that include separate matrices or dataframes of expression values for the potential causes and targets:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr(dX1::T, dX2::T, dG::T, dE::T; colG=1, colX=2, combination=\"IV\") where T<:AbstractDataFrame","category":"page"},{"location":"testLLR/#BioFindr.findr-Union{Tuple{T}, NTuple{4, T}} where T<:DataFrames.AbstractDataFrame","page":"Tests to evaluate","title":"BioFindr.findr","text":"findr(dX1::T, dX2::T, dG::T, dE::T; colG=1, colX=2, method=\"moments\", combination=\"IV\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr(Matrix(dX1), Matrix(dX2), Matrix(dG), pairGX2) when the inputs dX1, dX2, and dG are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target, (Posterior) Probability, and qvalue columns. When DataFrames are used, only combined posterior probabilities can be returned (combination=\"IV\" (default), \"mediation\", or \"orig\").\n\nThe numeric mapping between column indices in Matrix(dG) and Matrix(dX2) is obtained from these inputs using the getpairs function and the optional parameters:\n\ncolG - name or number of variant ID column in dE, default 1\ncolX - name or number of gene ID column in dE, default 2\nnamesX - names of a possible subset of columns in dX to be considered as potential causal regulators (default names(dX))\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX2) (sorted=false)\n\nSee also findr(::Matrix,::Array,::Array,::Matrix), combineprobs, stackprobs, globalfdr!.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"or with matrix-based input data:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr_matrix(X1::Matrix{T}, X2::Array{T}, G::Array{S}, pairGX::Matrix{R}; combination=\"none\")  where {T<:AbstractFloat, S<:Integer, R<:Integer}","category":"page"},{"location":"testLLR/#BioFindr.findr_matrix-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}, Matrix{R}}} where {T<:AbstractFloat, S<:Integer, R<:Integer}","page":"Tests to evaluate","title":"BioFindr.findr_matrix","text":"findr_matrix(X1::Matrix{T},X2::Array{T},G::Array{S},pairGX::Matrix{R}; method=\"moments\", combination=\"none\")  where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero causal relations from columns of input matrix X2 to columns of input matrix X1. The probabilities are estimated for a subset of columns of X2 that have a (discrete) instrumental variable in input matrix G. The matching between columns of X2 and columns of G is given by pairGX, a two-column array where the first column corresponds to a column index in G and the second to a column index in X2.\n\nPosterior probabilities are computed for the following tests\n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nwhich can be combined into the mediation test (P_2 P_3; combination=\"mediation\"), the instrumental variable or non-independence test (P_2 P_5; combination=\"IV\"), or BioFindr's original combination (frac12(P_2 P_5 + P_4); combination=\"orig\"). By default, individual probability matrices for all tests are returned (combination=\"none\").\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nIf combination=\"none\", then the output has size ncols(X1) x 4 x ncols(X2), where the middle index indexes the tests, and otherwise the output has size ncols(X1) x ncols(X2).\n\nSee also findr(::DataFrame,::DataFrame,::DataFrame,::DataFrame), combineprobs, supernormalize, pprob_col\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.\n\n\n\n\n\n","category":"method"},{"location":"testLLR/#Summary","page":"Tests to evaluate","title":"Summary","text":"","category":"section"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"A summary of all possible calls to the findr function:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr","category":"page"},{"location":"testLLR/#BioFindr.findr","page":"Tests to evaluate","title":"BioFindr.findr","text":"findr(dX::T; colnames=[], method=\"moments\", FDR=1.0, sorted=true, combination=\"none\") where T<:AbstractDataFrame\n\nWrapper for findr_matrix(Matrix(dX)) when the input dX is in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target, (Posterior) Probability, and qvalue columns. All columns of the input DataFrame must have a count or continuous scientific type. \n\nThe optional parameter colnames (vector of strings) determines whether we consider all columns of dX as source nodes (colnames=[], default), or only a subset of columns determined by the variable names in the vector colnames.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nThe optional parameter combination determines whether the output must be symmetrized. Possible values are none (default), prod, mean, or anti. If the optional parameter colnames is non-empty, symmetrization makes no sense and an error will be thrown unless combination=\"none\".\n\nSee also coerce_scitypes!, findr_matrix(::Matrix), symprobs, stackprobs, globalfdr!.\n\n\n\n\n\nfindr(dX::T, dG::T; method=\"moments\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr_matrix(Matrix(dX), Matrix(dG)) when the inputs dX and dG are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target (Posterior) Probability, and qvalue columns.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nNote that depending on the scitype of dG, different matrix-based methods are called. If the scitype dG is Count or Continuous, posterior probabilities for nonzero pairwise correlations between the variables in dG and variables in dX are computed. If the scitype of dG is Multiclass or OrderedFactor, posterior probabilities for nonzero differential expression of variables in dX across groups defined by the variables in dG are computed\n\nSee also coerce_scitypes!, findr_matrix(::Matrix,::Array), stackprobs, globalfdr!. \n\n\n\n\n\nfindr(dX::T, dG::T, dE::T; colX=2, colG=1, method=\"moments\", combination=\"IV\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr(Matrix(dX), Matrix(dG), pairGX) when the inputs are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target (Posterior) Probability, and qvalue columns. When DataFrames are used, only combined posterior probabilities can be returned (combination=\"IV\" (default), \"mediation\", or \"orig\").\n\nThe input dataframes are:\n\ndX - DataFrame with expression data, columns are genes\ndG - DataFrame with genotype data, columns are variants (SNPs)\ndE - DataFrame with eQTL results, must contain columns with gene and SNP IDs that can be mapped to column names in dX and dG, respectively\n\nThe numeric mapping between column indices in Matrix(dG) and Matrix(dX) is obtained from these inputs using the getpairs function and the optional parameters:\n\ncolG - name or number of variant ID column in dE, default 1\ncolX - name or number of gene ID column in dE, default 2\nnamesX - names of a possible subset of columns in dX to be considered as potential causal regulators (default names(dX))\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX) (sorted=false)\n\nSee also findr(::Matrix,::Array,::Matrix), getpairs, combineprobs, stackprobs, globalfdr!.\n\n\n\n\n\nfindr(dX1::T, dX2::T, dG::T, dE::T; colG=1, colX=2, method=\"moments\", combination=\"IV\", FDR=1.0, sorted=true) where T<:AbstractDataFrame\n\nWrapper for findr(Matrix(dX1), Matrix(dX2), Matrix(dG), pairGX2) when the inputs dX1, dX2, and dG are in the form of a DataFrame. The output is then also wrapped in a DataFrame with Source, Target, (Posterior) Probability, and qvalue columns. When DataFrames are used, only combined posterior probabilities can be returned (combination=\"IV\" (default), \"mediation\", or \"orig\").\n\nThe numeric mapping between column indices in Matrix(dG) and Matrix(dX2) is obtained from these inputs using the getpairs function and the optional parameters:\n\ncolG - name or number of variant ID column in dE, default 1\ncolX - name or number of gene ID column in dE, default 2\nnamesX - names of a possible subset of columns in dX to be considered as potential causal regulators (default names(dX))\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter FDR can be used to return only a subset of interactions with a desired expected FDR value (q-value threshold) (default 1.0, no filtering).\n\nThe optional parameter sorted determines if the output must be sorted by increasing q-value / decreasing posterior probability (sorted=true, the default) or by causal factor (column names of dX2) (sorted=false)\n\nSee also findr(::Matrix,::Array,::Array,::Matrix), combineprobs, stackprobs, globalfdr!.\n\n\n\n\n\n","category":"function"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"and the findr_matrix function:","category":"page"},{"location":"testLLR/","page":"Tests to evaluate","title":"Tests to evaluate","text":"findr_matrix","category":"page"},{"location":"testLLR/#BioFindr.findr_matrix","page":"Tests to evaluate","title":"BioFindr.findr_matrix","text":"findr_matrix(X::Matrix{T}; cols=[], method=\"moments\", combination=\"none\") where T<:AbstractFloat\n\nCompute posterior probabilities for nonzero pairwise correlations between columns of input matrix X. The probabilities are directed (asymmetric) in the sense that they are estimated from a column-specific background distribution.\n\nThe optional parameter cols (vector of integers) determines whether we consider all columns of X as source nodes (cols=[], default), or only a subset of columns determined by the indices in the vector cols.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nThe optional parameter combination determines whether the output must be symmetrized. Possible values are none (default), prod, mean, or anti. If the optional parameter cols is non-empty, symmetrization makes no sense and an error will be thrown unless combination=\"none\".\n\nSee also findr(::DataFrame), symprobs, supernormalize, pprob_col.\n\n\n\n\n\nfindr_matrix(X1::Matrix{T}, X2::Matrix{T}; method=\"moments\") where T<:AbstractFloat\n\nCompute posterior probabilities for nonzero pairwise correlations between columns of input matrices X1 and X2. The probabilities are directed (asymmetric) from the columns of X2 to the columns of X1 in the sense that they are estimated from a column-specific background distribution for each column of X2.\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nOnly use this method if X1 and X2 are distinct (no overlapping columns). For X2 consisting of a subset of columns with indices idx, use findr(X1; cols=idx) instead. \n\nSee also findr(::DataFrame), symprobs, supernormalize, pprob_col.\n\n\n\n\n\nfindr_matrix(X::Matrix{T},G::Array{S}; method=\"moments\") where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero differential expression of colunns of input matrix X across groups defined by one or more categorical variables (columns of G).\n\nReturn a matrix of size ncols(X) x ncols(G)\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nSee also findr(::DataFrame,::DataFrame), supernormalize, pprob_col.\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. CategoricalArrays will be supported in the future.\n\n\n\n\n\nfindr_matrix(X::Matrix{T},G::Matrix{S},pairGX::Matrix{S}; method=\"moments\", combination=\"none\") where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero causal relations between columns of input matrix X. The probabilities are estimated for relations going from a subset of columns of X that have a (discrete) instrumental variable in input matrix G to all columns of X, while excluding self-interactions (given default value 1). The matching between columns of X and columns of G is given by pairGX, a two-column array where the first column corresponds to a column index in G and the second to a column index in X.\n\nPosterior probabilities are computed for the following tests\n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nwhich can be combined into the mediation test (P_2 P_3; combination=\"mediation\"), the instrumental variable or non-independence test (P_2 P_5; combination=\"IV\"), or BioFindr's original combination (frac12(P_2 P_5 + P_4); combination=\"orig\"). By default, individual probability matrices for all tests are returned (combination=\"none\").\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nIf combination=\"none\", then the output has size ncols(X) x 4 x ncols(G), where the middle index indexes the tests, and otherwise the output has size ncols(X) x ncols(G). \n\nSee also findr(::DataFrame,::DataFrame,::DataFrame), supernormalize, pprob_col, combineprobs.\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.\n\n\n\n\n\nfindr_matrix(X1::Matrix{T},X2::Array{T},G::Array{S},pairGX::Matrix{R}; method=\"moments\", combination=\"none\")  where {T<:AbstractFloat, S<:Integer}\n\nCompute posterior probabilities for nonzero causal relations from columns of input matrix X2 to columns of input matrix X1. The probabilities are estimated for a subset of columns of X2 that have a (discrete) instrumental variable in input matrix G. The matching between columns of X2 and columns of G is given by pairGX, a two-column array where the first column corresponds to a column index in G and the second to a column index in X2.\n\nPosterior probabilities are computed for the following tests\n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nwhich can be combined into the mediation test (P_2 P_3; combination=\"mediation\"), the instrumental variable or non-independence test (P_2 P_5; combination=\"IV\"), or BioFindr's original combination (frac12(P_2 P_5 + P_4); combination=\"orig\"). By default, individual probability matrices for all tests are returned (combination=\"none\").\n\nThe optional parameter method determines the LLR mixture distribution fitting method and can be either moments (default) for the method of moments, or kde for kernel-based density estimation.\n\nIf combination=\"none\", then the output has size ncols(X1) x 4 x ncols(X2), where the middle index indexes the tests, and otherwise the output has size ncols(X1) x ncols(X2).\n\nSee also findr(::DataFrame,::DataFrame,::DataFrame,::DataFrame), combineprobs, supernormalize, pprob_col\n\nnote: Note\nG is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.\n\n\n\n\n\n","category":"function"},{"location":"listfunctions/","page":"List of functions","title":"List of functions","text":"CurrentModule = BioFindr","category":"page"},{"location":"listfunctions/#List-of-package-functions","page":"List of functions","title":"List of package functions","text":"","category":"section"},{"location":"listfunctions/","page":"List of functions","title":"List of functions","text":"","category":"page"},{"location":"bayesiannets/","page":"Bayesian networks","title":"Bayesian networks","text":"CurrentModule = BioFindr","category":"page"},{"location":"bayesiannets/#Bayesian-network-learning","page":"Bayesian networks","title":"Bayesian network learning","text":"","category":"section"},{"location":"bayesiannets/","page":"Bayesian networks","title":"Bayesian networks","text":"BioFindr implements the method described in the paper High-dimensional bayesian network inference from systems genetics data using genetic node ordering[Wang2019] to learn a Bayesian network using a dataframe of posterior probabilities as prior edge weights.","category":"page"},{"location":"bayesiannets/#DAG-reconstruction","page":"Bayesian networks","title":"DAG reconstruction","text":"","category":"section"},{"location":"bayesiannets/","page":"Bayesian networks","title":"Bayesian networks","text":"The first step in the algorithm is to convert a dataframe of posterior probabilities to a directed acyclic graph (DAG). [BioFindr][1] implements the original greedy algorithm where edges are added one by one in descending order of posterior probability and edges that would introduce a cycle in the dagfindr_greedy_edges! function. Two additional methods dagfindr_greedy_insertion! and dagfindr_heuristic_sort! developed by Kenneth Stoop and Pieter Audenaert in this paper are also implemented. The dagfindr! is the main user interface function.","category":"page"},{"location":"bayesiannets/","page":"Bayesian networks","title":"Bayesian networks","text":"dagfindr!\ndagfindr_greedy_edges!\ndagfindr_heuristic_sort!\ndagfindr_greedy_insertion!\ngreedy_insertions!\nedge_weights\nnames_to_index!","category":"page"},{"location":"bayesiannets/#BioFindr.dagfindr!","page":"Bayesian networks","title":"BioFindr.dagfindr!","text":"dagfindr!(dP::T; method=\"greedy edges\") where T<:AbstractDataFrame\n\nConvert a DataFrame dP of findr results (list of edges) to a directed acyclic graph (DAG) using the specified method. The output is a directed graph represented as a SimpleDiGraph from the Graphs package. The method can be any of\n\n\"greedy edges\" (default), see dagfindr_greedy_edges!.\n\"heuristic sort\", see dagfindr_heuristic_sort!.\n\"greedy insertion\", see dagfindr_greedy_insertion!.\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.dagfindr_greedy_edges!","page":"Bayesian networks","title":"BioFindr.dagfindr_greedy_edges!","text":"dagfindr_greedy_edges!(dP::T) where T<:AbstractDataFrame\n\nConvert a DataFrame of dP of findr results (list of edges) to a directed acyclic graph (DAG) represented as a SimpleDiGraph from the Graphs package. This function implements the method of Wang et al. (2019) where edges are added one-by-one in decreasing order of probability, and only if they do not create a cycle in the graph, using the incremental cycle detection algorithm from the Graphs package.\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.dagfindr_heuristic_sort!","page":"Bayesian networks","title":"BioFindr.dagfindr_heuristic_sort!","text":"dagfindr_heuristic_sort!(dP::T) where T<:AbstractDataFrame\n\nConvert a DataFrame of dP of findr results (list of edges) to a directed acyclic graph (DAG) represented as a SimpleDiGraph from the Graphs package. This function implements the heuristic sort method of Stoop et al. (2023) where vertices are sorted by their ratio of out-degree to in-degree, and edges are added only if their source vertex precedes their target vertex in the sorted list. The output is a directed graph and a dictionary to map vertex names to numbers.\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.dagfindr_greedy_insertion!","page":"Bayesian networks","title":"BioFindr.dagfindr_greedy_insertion!","text":"dagfindr_greedy_insertion!(dP::T) where T<:AbstractDataFrame\n\nConvert a DataFrame of dP of findr results (list of edges) to a directed acyclic graph (DAG) represented as a SimpleDiGraph from the Graphs package. This function implements the greedy insertion method of Stoop et al. (2023) where vertices are sorted iteratively by inserting vertices in the position in the current ordering yields the maximum possible gain of edge weights, where the gain is counted as the difference between the sum of new edges weight included and the sum of old edge weights lost, where edges are counted only if their source vertex precedes their target vertex in the ordering. The output is a directed graph and a dictionary to map vertex names to numbers.\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.greedy_insertions!","page":"Bayesian networks","title":"BioFindr.greedy_insertions!","text":"greedy_insertions!(sorted_vertices, weights)\n\nTBW\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.edge_weights","page":"Bayesian networks","title":"BioFindr.edge_weights","text":"edge_weights(dP::T) where T<:AbstractDataFrame\n\nTBW\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/#BioFindr.names_to_index!","page":"Bayesian networks","title":"BioFindr.names_to_index!","text":"names_to_index!(dP::T) where T<:AbstractDataFrame\n\nAdd columns with vertex numbers to a DataFrame dP of edges. The columns Source_idx and Target_idx are added to dP with the vertex numbers corresponding to the names in the Source and Target columns, respectively. The function returns a dictionary name2num to map vertex names to numbers.\n\n\n\n\n\n","category":"function"},{"location":"bayesiannets/","page":"Bayesian networks","title":"Bayesian networks","text":"[Wang2019]: Wang L, Audenaert P, Michoel T (2019) High-dimensional bayesian network inference from systems genetics data using genetic node ordering. Frontiers in Genetics, Special Topic Machine Learning and Network-Driven Integrative Genomics, 10, 1196.","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"CurrentModule = BioFindr","category":"page"},{"location":"#BioFindr-Documentation","page":"Introduction","title":"BioFindr Documentation","text":"","category":"section"},{"location":"","page":"Introduction","title":"Introduction","text":"This is the documentation for BioFindr, an implementation of the Findr software in Julia. ","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"The methods implemented in BioFindr were developed by Lingfei Wang and Tom Michoel, and were first described in the paper \"Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data\".[Wang2017]","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"This documentation copies both the structure and (most of) the contents of the Materials and methods section of the original paper, directly linking the mathematical description of the method to the documentation and source code of its implementation. Maybe this is how methods papers in computational biology should be written in the first place?","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"If you haven't used BioFindr before, the BioFindrTutorials website may be useful.","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"If you are familiar with the original Findr software, you should be aware that BioFindr.jl is not a literal translation. In particular, pay attention to the following differences:","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"The main findr interface function for Causal inference takes as input expression and genotype matrices or DataFrames, and a list or DataFrame of pairs to match eQTLs with a subset of genes. This avoids the need to reshape the gene expression data every time findr is called with a different set of eQTLs.\nInput and output are structured by columns, that is, in the gene expression and genotype data, columns are genes or SNPs and rows are samples, and in the posterior probability matrices, each column contains the probabilities of a causal relation from the gene corresponding to that column to all other genes. This is the opposite of the original software where variables corresponded to rows. This is to boost performance as Julia stores arrays in column-major format.\nIf you call findr with DataFrame inputs (where columns naturally correspond to variables, that is, genes or SNPs), you need not worry about remembering the role of rows and columns in the output, as the output is returned in the form of a DataFrame with Source, Target, Probability, and q-value columns.\nYou can pass a desired global FDR value for filtering inferred associations as a parameter when calling findr, no more need for manual post-processing of the output.  \nEstimation of the observed distribution of log-likelihood ratios uses either a new, parametric method of moments, or a kernel density estimation method, replacing the previous histogram-based method. ","category":"page"},{"location":"#Table-of-contents","page":"Introduction","title":"Table of contents","text":"","category":"section"},{"location":"","page":"Introduction","title":"Introduction","text":"","category":"page"},{"location":"","page":"Introduction","title":"Introduction","text":"[Wang2017]: Wang L, Michoel T (2017) Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data. PLoS Comput Biol 13(8): e1005703.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"CurrentModule = BioFindr","category":"page"},{"location":"inference/#General-inference-algorithm","page":"General inference algorithm","title":"General inference algorithm","text":"","category":"section"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"Consider a set of observations sampled from a mixture distribution of a null and an alternative hypothesis. For instance in gene regulation, every observation can correspond to expression levels of a pair of genes wich are sampled from a bivariate normal distribution with zero (null hypothesis) or non-zero (alternative hypothesis) correlation coefficient. In BioFindr, we predict the probability that any sample follows the alternative hypothesis with the following algorithm (based on and modified from [Chen2007]):","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"1.  For robustness against outliers, we convert every continuous variable into standard normally distributed N(01) values using a rank-based inverse normal transformation across all samples. We name this step as supernormalization.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"supernormalize","category":"page"},{"location":"inference/#BioFindr.supernormalize","page":"General inference algorithm","title":"BioFindr.supernormalize","text":"supernormalize(X[, c])\n\nConvert each column of matrix or dataframe X of reals into standard normally distributed values using a rank-based inverse normal transformation. Then scale each column to have variance one.\n\nNote that after the inverse normal transformation, each column has mean zero and identical variance (if we use ordinal ranking). Hence rescaling can be done once on the whole matrix.\n\nThe formula and default value for the paramater c come from this paper\n\n\n\n\n\n","category":"function"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"2.  We propose a null and an alternative hypothesis for all Likelihood ratio tests of interest where, by definition, the null hypothesis space is a subset of the alternative hypothesis. Model parameters are replaced with their maximum likelihood estimators (MLEs) to obtain the log likelihood ratio (LLR) between the alternative and null hypotheses.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"3.  We derive the analytical expression for the probablity density function (PDF) of the Null distributions of the log-likelihood ratios when samples follow the null hypothesis.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"4.  We convert LLRs using Bayesian inference of posterior probabilities of the hypothesis of interest with empirical estimation of local false discovery rate.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"5. We consider multiple Tests to evaluate, consisting of combinations of the basic Likelihood ratio tests, for common tasks in genome-wide studies: ","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"Coexpression analysis \nAssociation analysis\nCausal inference\nBipartite causal inference.","category":"page"},{"location":"inference/","page":"General inference algorithm","title":"General inference algorithm","text":"[Chen2007]: Chen L, Emmert-Streib F, Storey J. Harnessing naturally randomized transcription to infer regulatory relationships among genes. Genome Biol 8, R219 (2007).","category":"page"}]
}
