<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Introduction · BioFindr.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/BioFindr.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>BioFindr.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Introduction</a><ul class="internal"><li><a class="tocitem" href="#Table-of-contents"><span>Table of contents</span></a></li></ul></li><li><a class="tocitem" href="inference/">General inference algorithm</a></li><li><a class="tocitem" href="realLLR/">Likelihood ratio tests</a></li><li><a class="tocitem" href="randomLLR/">Null distributions of the LLRs</a></li><li><a class="tocitem" href="posteriorprobs/">Bayesian inference of posterior probabilities</a></li><li><a class="tocitem" href="testLLR/">Tests to evaluate</a></li><li><a class="tocitem" href="utils/">Utilities</a></li><li><a class="tocitem" href="listfunctions/">List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Introduction</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Introduction</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/BioFindr.jl/blob/main/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="BioFindr.jl-Documentation"><a class="docs-heading-anchor" href="#BioFindr.jl-Documentation">BioFindr.jl Documentation</a><a id="BioFindr.jl-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#BioFindr.jl-Documentation" title="Permalink"></a></h1><p>This is the documentation for <a href="https://github.com/tmichoel/BioFindr.jl">BioFindr.jl</a>, an implementation of the <a href="https://github.com/lingfeiwang/findr">Findr software</a> in <a href="https://julialang.org/">Julia</a>. </p><p>The methods implemented in BioFindr were developed by <a href="https://github.com/lingfeiwang">Lingfei Wang</a> and <a href="https://tmichoel.github.io">Tom Michoel</a>, and were first described in the paper <a href="https://doi.org/10.1371/journal.pcbi.1005703">&quot;Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data&quot;</a>.<sup class="footnote-reference"><a id="citeref-Wang2017" href="#footnote-Wang2017">[Wang2017]</a></sup></p><p>This documentation copies both the structure and (most of) the contents of the <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005703#sec009">Materials and methods section</a> of the original paper, directly linking the mathematical description of the method to the documentation and source code of its implementation. Maybe this is how methods papers in computational biology should be written in the first place?</p><p>If you haven&#39;t used BioFindr before, the <a href="https://tmichoel.github.io/BioFindrTutorials">BioFindrTutorials website</a> may be useful.</p><p>If you are familiar with the <a href="https://github.com/lingfeiwang/findr">original Findr software</a>, you should be aware that <a href="https://github.com/tmichoel/BioFindr.jl">BioFindr.jl</a> is not a literal translation. In particular, pay attention to the following differences:</p><ul><li><p>The main <code>findr</code> interface function for <a href="testLLR/#Causal-inference">Causal inference</a> takes as input expression and genotype matrices or <a href="https://dataframes.juliadata.org/stable/">DataFrames</a>, and a list or <a href="https://dataframes.juliadata.org/stable/">DataFrame</a> of pairs to match eQTLs with a subset of genes. This avoids the need to reshape the gene expression data every time <code>findr</code> is called with a different set of eQTLs.</p></li><li><p>Input and output are structured by <strong>columns</strong>, that is, in the gene expression and genotype data, columns are genes or SNPs and rows are samples, and in the posterior probability matrices, each column contains the probabilities of a causal relation from the gene corresponding to that column to all other genes. This is the <strong>opposite</strong> of the <a href="https://github.com/lingfeiwang/findr">original software</a> where variables corresponded to rows. This is to boost performance as Julia stores arrays in <a href="https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-column-major">column-major</a> format.</p></li><li><p>If you call <code>findr</code> with <a href="https://dataframes.juliadata.org/stable/">DataFrame</a> inputs (where columns naturally correspond to variables, that is, genes or SNPs), you need not worry about remembering the role of rows and columns in the output, as the output is returned in the form of a <a href="https://dataframes.juliadata.org/stable/">DataFrame</a> with <code>Source</code>, <code>Target</code>, <code>Probability</code>, and <code>q-value</code> columns.</p></li><li><p>You can pass a desired global FDR value for filtering inferred associations as a parameter when calling <code>findr</code>, no more need for manual post-processing of the output.  </p></li><li><p>Estimation of the observed distribution of log-likelihood ratios uses either a <a href="posteriorprobs/#mom_postprobs">new, parametric method of moments</a>, or a <a href="posteriorprobs/#kde_postprobs">kernel density estimation method</a>, replacing the previous histogram-based method. </p></li></ul><h2 id="Table-of-contents"><a class="docs-heading-anchor" href="#Table-of-contents">Table of contents</a><a id="Table-of-contents-1"></a><a class="docs-heading-anchor-permalink" href="#Table-of-contents" title="Permalink"></a></h2><ul><li><a href="#BioFindr.jl-Documentation">BioFindr.jl Documentation</a></li><li class="no-marker"><ul><li><a href="#Table-of-contents">Table of contents</a></li></ul></li><li><a href="inference/#General-inference-algorithm">General inference algorithm</a></li><li><a href="listfunctions/#List-of-package-functions">List of package functions</a></li><li><a href="posteriorprobs/#Bayesian-inference-of-posterior-probabilities">Bayesian inference of posterior probabilities</a></li><li class="no-marker"><ul><li><a href="posteriorprobs/#Estimating-P({\\mathcal-H}_{\\mathrm{null}})">Estimating <span>$P({\mathcal H}_{\mathrm{null}})$</span></a></li><li><a href="posteriorprobs/#mom_postprobs">Method of moments estimation of the mixture distribution</a></li><li><a href="posteriorprobs/#kde_postprobs">Kernel density estimation of the mixture distribution</a></li></ul></li><li><a href="randomLLR/#Null-distributions-of-the-log-likelihood-ratios">Null distributions of the log-likelihood ratios</a></li><li class="no-marker"><ul><li><a href="randomLLR/#Correlation-test">Correlation test</a></li><li><a href="randomLLR/#Primary-linkage-test">Primary linkage test</a></li><li><a href="randomLLR/#Secondary-linkage-test">Secondary linkage test</a></li><li><a href="randomLLR/#Conditional-independence-test">Conditional independence test</a></li><li><a href="randomLLR/#Relevance-test">Relevance test</a></li><li><a href="randomLLR/#Pleiotropy-test">Pleiotropy test</a></li><li><a href="randomLLR/#Implementation">Implementation</a></li><li><a href="randomLLR/#LBeta-related-functions">LBeta related functions</a></li></ul></li><li><a href="realLLR/#Likelihood-ratio-tests">Likelihood ratio tests</a></li><li class="no-marker"><ul><li><a href="realLLR/#corr_test_realLLR">Correlation test</a></li><li><a href="realLLR/#prim_test_realLLR">Primary linkage test</a></li><li><a href="realLLR/#Secondary-linkage-test">Secondary linkage test</a></li><li><a href="realLLR/#Conditional-independence-test">Conditional independence test</a></li><li><a href="realLLR/#Relevance-test">Relevance test</a></li><li><a href="realLLR/#Controlled-or-pleiotropy-test">Controlled or pleiotropy test</a></li><li><a href="realLLR/#Implementation">Implementation</a></li></ul></li><li><a href="testLLR/#Tests-to-evaluate">Tests to evaluate</a></li><li class="no-marker"><ul><li><a href="testLLR/#Coexpression-analysis">Coexpression analysis</a></li><li><a href="testLLR/#Association-analysis">Association analysis</a></li><li><a href="testLLR/#Causal-inference">Causal inference</a></li><li><a href="testLLR/#Bipartite-causal-inference">Bipartite causal inference</a></li><li><a href="testLLR/#Summary">Summary</a></li></ul></li><li><a href="utils/#Utilities">Utilities</a></li><li><a href="utils/#Generating-simulated-data">Generating simulated data</a></li></ul><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-Wang2017"><a class="tag is-link" href="#citeref-Wang2017">Wang2017</a>Wang L, Michoel T (2017) <a href="https://doi.org/10.1371/journal.pcbi.1005703">Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data</a>. PLoS Comput Biol 13(8): e1005703.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="inference/">General inference algorithm »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Friday 5 April 2024 20:26">Friday 5 April 2024</span>. Using Julia version 1.10.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
