var documenterSearchIndex = {"docs":
[{"location":"","page":"Home","title":"Home","text":"CurrentModule = Findr","category":"page"},{"location":"#Findr","page":"Home","title":"Findr","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Documentation for Findr.","category":"page"},{"location":"","page":"Home","title":"Home","text":"","category":"page"},{"location":"","page":"Home","title":"Home","text":"Modules = [Findr]","category":"page"},{"location":"#Distributions.ccdf-Tuple{Findr.LBeta, Real}","page":"Home","title":"Distributions.ccdf","text":"ccdf(d,x)\n\nEvaluate the complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"method"},{"location":"#Distributions.cdf-Tuple{Findr.LBeta, Real}","page":"Home","title":"Distributions.cdf","text":"cdf(d,x)\n\nEvaluate the cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"method"},{"location":"#Distributions.fit_mle-Union{Tuple{T}, Tuple{Type{<:Findr.LBeta}, AbstractArray{T}}} where T<:Real","page":"Home","title":"Distributions.fit_mle","text":"fit_mle(LBeta, x)\n\nFit an LBeta distribution to data x using maximum-likelihood estimation by exploiting its relationship to the Beta distribution.\n\nSee https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl\n\n\n\n\n\n","category":"method"},{"location":"#Distributions.logccdf-Tuple{Findr.LBeta, Real}","page":"Home","title":"Distributions.logccdf","text":"logccdf(d,x)\n\nEvaluate the complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. \n\n\n\n\n\n","category":"method"},{"location":"#Distributions.pdf-Tuple{Findr.LBeta, Real}","page":"Home","title":"Distributions.pdf","text":"pdf(d,x)\n\nEvaluate the probability density function of an LBeta distribution with support on xgeq 0. \n\n\n\n\n\n","category":"method"},{"location":"#Findr.findr_causal-Tuple{Any, Any, Any}","page":"Home","title":"Findr.findr_causal","text":"findr_causal(X,E,pairEX)\n\nCompute posterior probabilities for nonzero causal relations between rows of input matrix X. The probabilities are estimated for a subset of rows of X that have a (discrete) instrumental variable in input matrix E. The matching between rows of X and rows of E is given by pairEX, a two-column array where the first column corresponds to a row index in E and the second to a row index in X.\n\nPosterior probabilities are computed for the following tests\n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nwhich can be combined into the mediation test (P2P3), the non-independence test (P2P5), or Findr's legacy combination (0.5(P2P5 + P4)). Alternatively, individual probability matrices for all tests are returned.\n\nAll return matrices have size numrows(E) x numrows(X).\n\n\n\n\n\n","category":"method"},{"location":"#Findr.findr_corr-Tuple{Any}","page":"Home","title":"Findr.findr_corr","text":"findr_corr(X)\n\nCompute posterior probabilities for nonzero pairwise correlations between rows of input matrix X. The probabilities are directed in the sense that they are estimated from a row-specific background distribution.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.findr_diffexp-Tuple{Any, Any}","page":"Home","title":"Findr.findr_diffexp","text":"findr_diffexp(X,E)\n\nCompute posterior probabilities for nonzero differential expression of rows of input matrix X across groups defined by one or more categorical variables (rows of E).\n\nReturn a matrix of size numrows(E) x numrows(X)\n\n\n\n\n\n","category":"method"},{"location":"#Findr.fit_kde-Tuple{Any}","page":"Home","title":"Findr.fit_kde","text":"fit_kde(llr)\n\nFit a distribution function to a vector of log-likelihood ratio values llr using kernel density estimation. To avoid boundary effects in the KDE, the log-likelihoods (which take values in 0infty)) are first transformed to a vector of z,\n\nz = log left( e^2 LLR - 1 right)\n\nwhich takes values in (-inftyinfty). KDE is applied to z, and the probability density function (pdf) for llr is obtained from the pdf for z by the usual transformation rule for functions of random variables.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.fit_mixdist_EM","page":"Home","title":"Findr.fit_mixdist_EM","text":"fit_mixdist_EM(llr,ns,ng=1,test=:corr; maxiter=1000, tol=1e-14)\n\nFit a two-component mixture distribution of two LBeta distribution to a vector of log-likelihood ratios llr using an EM algorithm. The first component is the true null distribution for a given Findr test with sample size ns and number of genotype groups ng. The second component is the alternative distribution, assumed to follow an LBeta distribution. The prior probability pi0 of an observation belonging to the null component is fixed and determined by the pi0est function. Hence only the parameters of the alternative component need to be estimated.\n\nThe EM algorithm outputs posterior probabilities of the alternative hypothesis being true, in the form of the estimated recognition distribution. The optional parameters maxiter (default value 1000) and tol (default value 1e-14) control the convergence of the EM algorithm.\n\nThe input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \":corr\", the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"#Findr.fit_mixdist_KDE","page":"Home","title":"Findr.fit_mixdist_KDE","text":"fit_mixdist_KDE(llr,ns,[ng,test])\n\nReturn posterior probabilities for a vector of log-likelihood ratio values llr for a given Findr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \":corr\", the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"#Findr.fit_weighted-Union{Tuple{T}, Tuple{Type{<:Findr.LBeta}, AbstractArray{T}, StatsBase.AbstractWeights}} where T<:Real","page":"Home","title":"Findr.fit_weighted","text":"fit_weighted(LBeta, x, w)\n\nFit an LBeta distribution to data x where each observation in x has a weight w between 0 and 1, using the method of moments. The function exploits the relationship to the Beta distribution and is a simple adaptation of the fit function in\n\nhttps://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl\n\n\n\n\n\n","category":"method"},{"location":"#Findr.groupmeans-Tuple{Any, Any}","page":"Home","title":"Findr.groupmeans","text":"groupmeans(Y,E)\n\nCompute the size and mean of each row of matrix Y for each of the groups (unique values) in categorical vector E.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.llrstats_row-Tuple{Any, Any, Any}","page":"Home","title":"Findr.llrstats_row","text":"llrstats_row(Y,E,row)\n\nCompute the sufficient statistics to compute the log-likelihood ratios for Findr tests 2-5  for a given row (gene) of gene expression matrix Y with categorical instrument E against all other rows of Y.\n\nY is assumed to have undergone supernormalization with each row having mean zero and variance one. The LLRs are scaled by the number of samples.\n\nThe sufficient statistics are:\n\nthe covariance ρ between the given row of matrix Y and all other rows of Y\nthe weighted average variances σ1 of each row of matrix Y over the groups (unique values) in E\nthe weighted average covariance σ2 between the given row of Y and all other rows of Y over the groups (unique values) of E\n\n\n\n\n\n","category":"method"},{"location":"#Findr.nulldist","page":"Home","title":"Findr.nulldist","text":"nulldist(ns,[ng,test])\n\nReturn an LBeta distribution that is the null distribution of the log-likelihood ratio for a given Findr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith only one input argument, the null distribution for the correlation test with ns samples is returned. With two input arguments, or  with three arguments and test equal to \"corr\", the null distribution for the correlation test with ns samples is returned and the second argument is ignored\n\n\n\n\n\n","category":"function"},{"location":"#Findr.nulllog10pval","page":"Home","title":"Findr.nulllog10pval","text":"nulllog10pval(llr,ns,[ng,test])\n\nReturn negative log10 p-values for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given Findr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \"corr\", the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"#Findr.nullpdf","page":"Home","title":"Findr.nullpdf","text":"nullpdf(llr,ns,[ng,test])\n\nReturn probability distribution function evaluations for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given Findr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \":corr\", the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"#Findr.nullpval","page":"Home","title":"Findr.nullpval","text":"nullpval(llr,ns,[ng,test])\n\nReturn p-values for a vector of log-likelihood ratio values llr under the null distribution of the log-likelihood ratio for a given Findr test with sample size ns and number of genotype groups ng. The input variable test can take the values:\n\n':corr' - correlation test (test 0)\n':link' - linkage test (test 1/2)\n':med' - mediation test (test 3)\n':relev' - relevance test (test 4)\n':pleio' - pleiotropy test (test 5)\n\nWith two input arguments, the correlation test with ns samples is used. With three input arguments, or with four arguments and test equal to \":corr\", the correlation test with ns samples is used and the third argument is ignored.\n\n\n\n\n\n","category":"function"},{"location":"#Findr.pi0est-Tuple{Any}","page":"Home","title":"Findr.pi0est","text":"pi0est(pval)\n\nEstimate the proportion π0 of truly null features in a vector pval of p-values using Storey's method\n\nSee also http://varianceexplained.org/files/pi0boot.pdf\n\n\n\n\n\n","category":"method"},{"location":"#Findr.pprob_causal_row-Tuple{Any, Any, Any}","page":"Home","title":"Findr.pprob_causal_row","text":"pprob_causal_row(Y,row)\n\nCompute the posterior probabilities for a given row (gene) of gene expression matrix Y with categorical instrument E against all other rows of Y for Findr causal tests: \n\n- Test 2 (**Linkage test**) \n- Test 3 (**Mediation test**)\n- Test 4 (**Relevance test**)\n- Test 5 (**Pleiotropy test**)\n\nY is assumed to have undergone supernormalization with each row having mean zero and variance one.\n\nFor test 2, 4, and 5 the posterior probabilities are the probabilities of the alternative hypothesis being true. For test 3 they are the probabilities of the null hypothesis being true.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.pprob_corr_row-Tuple{Any, Any}","page":"Home","title":"Findr.pprob_corr_row","text":"pprob_corr_row(Y,row)\n\nCompute the posterior probabilities for Findr test 0 (correlation test) for a given row (gene) of gene expression matrix Y against all other rows of Y.\n\nY is assumed to have undergone supernormalization with each row having mean zero and variance one. The LLRs are scaled by the number of samples.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.realLLRcausal_row-Tuple{Any, Any, Any}","page":"Home","title":"Findr.realLLRcausal_row","text":"realLLRcausal_row(Y,E,row)\n\nCompute for a given row (gene) of gene expression matrix Y with categorical instrument E against all other rows of Y the log-likelihood ratios for Findr causal tests: \n\nTest 2 (Linkage test) \nTest 3 (Mediation test)\nTest 4 (Relevance test)\nTest 5 (Pleiotropy test)\n\nY is assumed to have undergone supernormalization with each row having mean zero and variance one. The LLRs are scaled by the number of samples.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.realLLRcorr_row-Tuple{Any, Any}","page":"Home","title":"Findr.realLLRcorr_row","text":"realLLRcorr_row(Y,row)\n\nCompute the log-likelihood ratios for Findr test 0 (correlation test) for a given row (gene) of gene expression matrix Y against all other rows of Y.\n\nY is assumed to have undergone supernormalization with each row having mean zero and variance one. The LLRs are scaled by the number of samples.\n\n\n\n\n\n","category":"method"},{"location":"#Findr.supernormalize","page":"Home","title":"Findr.supernormalize","text":"supernormalize(X[, c])\n\nConvert each row of matrix X of reals into standard normally distributed values using a rank-based inverse normal transformation. Then scale each row to have variance one.\n\nNote that after the inverse normal transformation, each row has mean zero and identical variance (if we use ordinal ranking). Hence rescaling can be done once on the whole matrix.\n\nThe formula and default value for the paramater c come from this paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2921808/\n\n\n\n\n\n","category":"function"},{"location":"#StatsAPI.fit-Union{Tuple{T}, Tuple{Type{<:Findr.LBeta}, AbstractArray{T}}} where T<:Real","page":"Home","title":"StatsAPI.fit","text":"fit(LBeta, x)\n\nFit an LBeta distribution to data x using the method of moments by exploiting its relationship to the Beta distribution.\n\nSee https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl\n\n\n\n\n\n","category":"method"},{"location":"#StatsAPI.params-Tuple{Findr.LBeta}","page":"Home","title":"StatsAPI.params","text":"params(d)\n\nGet the parameters of an LBeta distribution.\n\n\n\n\n\n","category":"method"}]
}
