<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Bayesian inference of posterior probabilities · BioFindr.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/BioFindr.jl/posteriorprobs/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioFindr.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li><a class="tocitem" href="../inference/">General inference algorithm</a></li><li><a class="tocitem" href="../realLLR/">Likelihood ratio tests</a></li><li><a class="tocitem" href="../randomLLR/">Null distributions of the LLRs</a></li><li class="is-active"><a class="tocitem" href>Bayesian inference of posterior probabilities</a><ul class="internal"><li><a class="tocitem" href="#Estimating-P({\\mathcal-H}_{\\mathrm{null}})"><span>Estimating <span>$P({\mathcal H}_{\mathrm{null}})$</span></span></a></li><li><a class="tocitem" href="#mom_postprobs"><span>Method of moments estimation of the mixture distribution</span></a></li><li><a class="tocitem" href="#kde_postprobs"><span>Kernel density estimation of the mixture distribution</span></a></li></ul></li><li><a class="tocitem" href="../testLLR/">Tests to evaluate</a></li><li><a class="tocitem" href="../utils/">Utilities</a></li><li><a class="tocitem" href="../listfunctions/">List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Bayesian inference of posterior probabilities</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Bayesian inference of posterior probabilities</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/BioFindr.jl/blob/main/docs/src/posteriorprobs.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Bayesian-inference-of-posterior-probabilities"><a class="docs-heading-anchor" href="#Bayesian-inference-of-posterior-probabilities">Bayesian inference of posterior probabilities</a><a id="Bayesian-inference-of-posterior-probabilities-1"></a><a class="docs-heading-anchor-permalink" href="#Bayesian-inference-of-posterior-probabilities" title="Permalink"></a></h1><p>After obtaining the PDFs for the LLRs from real data and the null hypotheses, we can convert LLR values into posterior probabilities <span>$P({\mathcal H}_{\mathrm{alt}}\mid \mathrm{LLR})$</span>. We use a similar technique as in <sup class="footnote-reference"><a id="citeref-Chen2007" href="#footnote-Chen2007">[Chen2007]</a></sup>, which itself was based on a more general framework to estimate local FDRs in genome-wide studies <sup class="footnote-reference"><a id="citeref-Storey2003" href="#footnote-Storey2003">[Storey2003]</a></sup>. This framework assumes that the real distribution of a certain test statistic forms a mixture distribution of null and alternative hypotheses. After estimating the null distribution, either analytically or by simulation, it can be compared against the real distribution to determine the proportion of null hypotheses, and consequently the posterior probability that the alternative hypothesis is true at any value of the statistic.</p><p>To be precise, consider an arbitrary likelihood ratio test. The fundamental assumption is that in the limit <span>$\mathrm{LLR}\rightarrow 0^+$</span>, all test cases come from the null hypothesis (<span>${\mathcal H}_{\mathrm{null}}$</span>), whilst as <span>$\mathrm{LLR}$</span> increases, the proportion of alternative hypotheses (<span>${\mathcal H}_{\mathrm{alt}}$</span>) also grows. The mixture distribution of real LLR values is assumed to have a PDF as</p><p class="math-container">\[p(\mathrm{LLR}) = P({\mathcal H}_{\mathrm{null}}) p(\mathrm{LLR}\mid{\mathcal H}_{\mathrm{null}}) + P({\mathcal H}_{\mathrm{alt}})p(\mathrm{LLR}\mid{\mathcal H}_{\mathrm{alt}}).\]</p><p>The priors <span>$P({\mathcal H}_{\mathrm{null}})$</span> and <span>$P({\mathcal H}_{\mathrm{alt}})$</span> sum to unity and correspond to the proportions of null and alternative hypotheses in the mixture distribution. For any test <span>$i=0,\dots,5$</span>, Bayes&#39; theorem then yields its posterior probability as</p><p class="math-container">\[\begin{aligned}
P({\mathcal H}_{\mathrm{alt}}^{(i)}\mid\mathrm{LLR}^{(i)}) 
&amp;= \frac{p(\mathrm{LLR}^{(i)}\mid{\mathcal H}_{\mathrm{alt}}^{(i)})}{p(\mathrm{LLR}^{(i)})}P({\mathcal H}_{\mathrm{alt}}^{(i)})\\
&amp;= 1 - \frac{p(\mathrm{LLR}^{(i)}\mid{\mathcal H}_{\mathrm{null}}^{(i)})}{p(\mathrm{LLR}^{(i)})}P({\mathcal H}_{\mathrm{null}}^{(i)})
\end{aligned}\]</p><p>Based on this, we can define the posterior probabilities of the selected hypotheses according to the table in the <a href="../realLLR/#Likelihood-ratio-tests">Likelihood ratio tests</a> section, i.e. the alternative for tests 0, 1, 2, 4, 5 and the null for test 3 as</p><p class="math-container">\[P_i\equiv \begin{cases}
    P({\mathcal H}_{\mathrm{alt}}^{(i)}\mid\mathrm{LLR}^{(i)}), &amp; i=0,1,2,4,5\\
    P({\mathcal H}_{\mathrm{null}}^{(i)}\mid\mathrm{LLR}^{(i)}),&amp;i=3
\end{cases}\]</p><p>Distributions can be estimated either separately for every <span>$(E,A)$</span> pair or by pooling across all <span>$(E,A)$</span> pairs. In practice, we test on the order of <span>$10^3$</span> to <span>$10^4$</span> candidate targets (&quot;<span>$B$</span>&quot;) for every <span>$(E,A)$</span> such that a separate conversion of LLR values to posterior probabilities is both feasible and recommended, as it accounts for different roles of every gene, especially hub genes, through different rates of alternative hypotheses.</p><p>Lastly, in a typical application of BioFindr, inputs of <span>$(E,A)$</span> pairs will have been pre-determined as the set of significant eQTL-gene pairs from a genome-wide eQTL associaton analysis. In such cases, we may naturally assume <span>$P_1=1$</span> for all considered pairs, and skip the primary test.</p><p>Posterior probabilities are computed in the <a href="#BioFindr.pprob_col"><code>pprob_col</code></a> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.pprob_col" href="#BioFindr.pprob_col"><code>BioFindr.pprob_col</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">pprob_col(Y::Matrix{T},Ycol::Vector{T}; method=&quot;moments&quot;) where T&lt;:AbstractFloat</code></pre><p>Compute the posterior probabilities for BioFindr test 0 (<strong>correlation test</strong>) for a given column vector <code>Ycol</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>The optional parameter <code>method</code> determines the mixture distribution fitting method and can be either <code>moments</code> (default) for the method of moments, or <code>kde</code> for kernel-based density estimation.</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.fit_mixdist_mom"><code>fit_mixdist_mom</code></a>, <a href="#BioFindr.fit_mixdist_KDE"><code>fit_mixdist_KDE</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L1-L11">source</a></section><section><div><pre><code class="nohighlight hljs">pprob_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}; method=&quot;moments&quot;) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the posterior probabilities for the BioFindr causal tests for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of matrix <code>Y</code>: </p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one.</p><p>For test 2, 4, and 5 the posterior probabilities are the probabilities of the alternative hypothesis being true. For test 3 they are the probabilities of the null hypothesis being true.</p><p>The optional parameter <code>method</code> determines the mixture distribution fitting method and can be either <code>moments</code> (default) for the method of moments, or <code>kde</code> for kernel-based density estimation.</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.fit_mixdist_mom"><code>fit_mixdist_mom</code></a>, <a href="#BioFindr.fit_mixdist_KDE"><code>fit_mixdist_KDE</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L35-L52">source</a></section><section><div><pre><code class="nohighlight hljs">pprob_col(Y::Matrix{T},E::Vector{S}; method=&quot;moments&quot;) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the posterior probabilities for differential expression of columns of matrix <code>Y</code> in the groups defined by  categorical vector <code>E</code> using BioFindr test 2 (<strong>Linkage test</strong>) </p><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one.</p><p>The optional parameter <code>method</code> determines the mixture distribution fitting method and can be either <code>moments</code> (default) for the method of moments, or <code>kde</code> for kernel-based density estimation.</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.fit_mixdist_mom"><code>fit_mixdist_mom</code></a>, <a href="#BioFindr.fit_mixdist_KDE"><code>fit_mixdist_KDE</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L114-L124">source</a></section></article><p>Computation of the posterior probabilities requires estimation of the prior probability <span>$P({\mathcal H}_{\mathrm{null}})$</span> and fitting the mixture distribution <span>$p(\mathrm{LLR})$</span> from a set of random samples <span>$x_1, x_2,\dots,x_m$</span>.</p><h2 id="Estimating-P({\\mathcal-H}_{\\mathrm{null}})"><a class="docs-heading-anchor" href="#Estimating-P({\\mathcal-H}_{\\mathrm{null}})">Estimating <span>$P({\mathcal H}_{\mathrm{null}})$</span></a><a id="Estimating-P({\\mathcal-H}_{\\mathrm{null}})-1"></a><a class="docs-heading-anchor-permalink" href="#Estimating-P({\\mathcal-H}_{\\mathrm{null}})" title="Permalink"></a></h2><p>To estimate <span>$\pi_0\equiv P({\mathcal H}_{\mathrm{null}})$</span> we convert the samples <span>$x_1, x_2,\dots,x_m$</span> to p-values under the null hypothesis,</p><p class="math-container">\[p_i = P(X\geq x_i \mid {\mathcal H}_{\mathrm{null}}) = \int_{x_i}^\infty p(x \mid \alpha_0,\beta_0) = \mathrm{ccdf}(x_i\mid \alpha_0,\beta_0) \]</p><p>where <span>$\mathrm{ccdf}$</span> denotes the complementary CDF. The histogram of p-values would show the <a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/">characteristic shape of a set of anti-conservative p-values</a> and <span>$\pi_0$</span> can be estimated by a <a href="http://varianceexplained.org/files/pi0boot.pdf">robust bootstrap procedure</a>:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.pi0est" href="#BioFindr.pi0est"><code>BioFindr.pi0est</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">pi0est(pval)</code></pre><p>Estimate the proportion <span>$\pi_0$</span> of truly null features in a vector <code>pval</code> of p-values using a <a href="http://varianceexplained.org/files/pi0boot.pdf">bootstrap method</a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L283-L287">source</a></section></article><h2 id="mom_postprobs"><a class="docs-heading-anchor" href="#mom_postprobs">Method of moments estimation of the mixture distribution</a><a id="mom_postprobs-1"></a><a class="docs-heading-anchor-permalink" href="#mom_postprobs" title="Permalink"></a></h2><p>To fit the mixture distribution, the original method<sup class="footnote-reference"><a id="citeref-Wang2017" href="#footnote-Wang2017">[Wang2017]</a></sup> approximated PDFs with histograms. This requires proper choices of histogram bin widths, <span>$P({\mathcal H}_{\mathrm{null}})$</span>, and techniques to ensure the conversion from LLR to posterior probability is monotonically increasing and smooth.</p><p><a href="https://github.com/tmichoel/BioFindr.jl">BioFindr.jl</a> uses an alternative, parametric approach, where the basic assumption is that the alternative distribution <span>$p(\mathrm{LLR}\mid{\mathcal H}_{\mathrm{alt}})$</span> for each test also follows an <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution, that is, belongs to the same 2-parameter family of distributions as the null distributions. </p><p>Let us denote the random variable <span>$X=\mathrm{LLR}$</span> taking values <span>$x\geq 0$</span>, and <span>$\pi_0=P({\mathcal H}_{\mathrm{null}})$</span>. Our assumption is that</p><p class="math-container">\[X \sim \pi_0 \mathcal{D}(\alpha_0,\beta_0) + (1-\pi_0) \mathcal{D}(\alpha,\beta)\]</p><p>or</p><p class="math-container">\[p(x) = \pi_0 p(x \mid \alpha_0,\beta_0) + (1-\pi_0) p(x\mid \alpha,\beta)\]</p><p>where <span>$\mathcal{D}$</span> denotes an <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution with its PDF  <span>$p(x\mid \alpha,\beta)$</span> defined in <a href="../randomLLR/#Null-distributions-of-the-log-likelihood-ratios">Null distributions of the log-likelihood ratios</a>. </p><p>Since the parameters <span>$\alpha_0$</span> and <span>$\beta_0$</span> of the null distribution are know exactly (see <a href="../randomLLR/#Null-distributions-of-the-log-likelihood-ratios">Null distributions of the log-likelihood ratios</a>) and <span>$\pi_0$</span> can be estimated from the null p-values (see <a href="#BioFindr.pi0est"><code>pi0est</code></a>), it remains to estimate <span>$\alpha$</span>, and <span>$\beta$</span> from a set of random samples <span>$x_1, x_2,\dots,x_m$</span> of <span>$X$</span>.</p><p>To estimate <span>$\alpha$</span> and <span>$\beta$</span> we make use of the relation between the <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> and Beta distributions (see <a href="../randomLLR/#Null-distributions-of-the-log-likelihood-ratios">Null distributions of the log-likelihood ratios</a>). Define the random variable</p><p class="math-container">\[Y = 1 - e^{-2 X}\]</p><p>Since by assumption each component of the mixture distribution of <span>$X$</span> follows an <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution, it follows that <span>$Y$</span> follows a mixture of Beta distributions,</p><p class="math-container">\[Y \sim \pi_0 B\Bigl(\frac{\alpha_0}{2},\frac{\beta_0}{2}\Bigr) + (1 - \pi_0) B\Bigl(\frac{\alpha}{2},\frac{\beta}{2}\Bigr)\]</p><p>This result can be derived as follows. The function <span>$g(x)= 1-e^{-2x}$</span> increases monotonically for <span>$x\geq 0$</span> and has the inverse <span>$g^{-1}(y)=-\frac{1}{2}\ln(1-y)$</span>. By the usual rules of transforming random variables:</p><p class="math-container">\[\begin{aligned}
    P(Y\leq y) &amp;= P(1 - e^{-2 X} \leq y) \\
    &amp;= P(X \leq g^{-1}(y))\\
    &amp;= \pi_0 P(X \leq g^{-1}(y) \mid \alpha_0,\beta_0) + (1-\pi_0) P(X \leq g^{-1}(y) \mid \alpha,\beta)\\
    &amp;= \pi_0 F_0(g^{-1}(y)) + (1-\pi_0) F_1(g^{-1}(y))\\
    &amp;= \pi_0 G_0(y) + (1-\pi_0) G_1(y)
\end{aligned}\]</p><p>where <span>$F_0$</span> and <span>$F_1$</span> are the CDFs of random variables with distributions <span>$\mathcal{D}(\alpha_0,\beta_0)$</span> and <span>$\mathcal{D}(\alpha,\beta)$</span>, respectively, and <span>$G_0$</span> and <span>$G_1$</span> are the CDFs of random variables with distributions <span>$B(\alpha_0/2,\beta_0/2)$</span> and <span>$B(\alpha/2,\beta/2)$</span>. The last step used the transformation between <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> and Beta distributions derived in  <a href="../randomLLR/#Null-distributions-of-the-log-likelihood-ratios">Null distributions of the log-likelihood ratios</a>.</p><p>The parameters of a Beta distribution relate to <a href="https://en.wikipedia.org/wiki/Beta_distribution#Higher_moments">its first and second moment</a> as follows: if <span>$Z\sim B(a,b)$</span>, then</p><p class="math-container">\[\begin{aligned}
    \mathbb{E}(Z) &amp;= \frac{a}{a+b} \equiv m_1(a,b)\\
    \mathbb{E}(Z^2) &amp;= \mathbb{E}(Z) \frac{a+1}{a+b+1} \equiv m_2(a,b)
\end{aligned}\]</p><p>Conversely, if first and second moments <span>$m_1$</span> and <span>$m_2$</span> are given, these equations can be inverted to give the parameters</p><p class="math-container">\[\begin{aligned}
    a &amp;= \frac{m_1(m_1-m_2)}{(m_2-m_1^2)}\\
    b &amp;= \frac{(1-m_1)(m_1-m_2)}{(m_2-m_1^2)},
\end{aligned}\]</p><p>which requires <span>$m_1&gt;m_2&gt;m_1^2$</span></p><p>Multiplying by two then gives the parameters of an <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution given the moments of the corresponding Beta distribution:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.fit_mom" href="#BioFindr.fit_mom"><code>BioFindr.fit_mom</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mom(LBeta, m1, m2)</code></pre><p>Fit an <code>LBeta</code> distribution to given first and second moments <code>m1</code> and <code>m2</code> of the corresponding Beta distribution. This requires that <code>m1</code> and <code>m2</code> satisfy the following relations (see also the <a href="https://en.wikipedia.org/wiki/Beta_distribution#Two_unknown_parameters">Beta distribution wiki</a>):</p><p><span>$m_1&gt;m_2 \wedge m_2&gt;m_1^2$</span></p><p>An <a href="https://docs.julialang.org/en/v1/base/base/#Core.AssertionError">AssertionError</a> is thrown if the condition evaluates to <code>false</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/lbeta.jl#L107-L115">source</a></section></article><p>For the mixture distributed <span>$Y$</span>, we have</p><p class="math-container">\[\mathbb{E}(Y^k) = \pi_0 \mathbb{E}(Y^k \mid \alpha_0,\beta_0) + (1-\pi_0) \mathbb{E}(Y^k \mid \alpha,\beta)\]</p><p>and hence</p><p class="math-container">\[\begin{aligned}
    \mathbb{E}(Y) &amp;= \pi_0 m_1\bigl(\alpha_0/2,\beta_0/2\bigr) + (1-\pi_0) m_1\bigl(\alpha/2,\beta/2\bigr)\\
    \mathbb{E}(Y^2) &amp;= \pi_0 m_2\bigl(\alpha_0/2,\beta_0/2\bigr) + (1-\pi_0) m_2\bigl(\alpha/2,\beta/2\bigr)
\end{aligned}\]</p><p>or</p><p class="math-container">\[\begin{aligned}
    m_1\bigl(\alpha/2,\beta/2\bigr) &amp;= \frac{\mathbb{E}(Y) - \pi_0 m_1\bigl(\alpha_0/2,\beta_0/2\bigr)}{(1-\pi_0)}\\
    m_2\bigl(\alpha/2,\beta/2\bigr) &amp;= \frac{\mathbb{E}(Y^2) - \pi_0 m_2\bigl(\alpha_0/2,\beta_0/2\bigr)}{(1-\pi_0)}
\end{aligned}\]</p><p>Given samples <span>$x_1, x_2,\dots,x_m$</span> we transform them to</p><p class="math-container">\[y_i = 1 - e^{-2 x_i}\]</p><p>and replace <span>$\mathbb{E}(Y)$</span> and <span>$\mathbb{E}(Y^2)$</span> in these equations by their estimates <span>$\frac{1}{n}\sum_i y_i$</span> and <span>$\frac{1}{n}\sum_i y_i^2$</span>, respectively. Likewise we replace <span>$\pi_0$</span> by its estimate <span>$\hat{\pi}_0$</span> obtained as explained above (see <a href="#BioFindr.pi0est"><code>pi0est</code></a>). Since <span>$\alpha_0$</span> and <span>$\beta_0$</span> are known exactly, we obtain estimates <span>$\hat{m}_1$</span> and <span>$\hat{m}_2$</span> for the moments of a Beta distribution with parameters <span>$\alpha/2$</span> and <span>$\beta/2$</span>. Plugging these moment estimates in the <a href="#BioFindr.fit_mom"><code>fit_mom</code></a> function gives estimates <span>$\hat{\alpha}$</span> and <span>$\hat{\beta}$</span> for the corresponding <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution.</p><p>To make this estimated distribution a valid alternative distribution component of the LLRs mixture distribution, we need to ensure the validity of two edge cases:</p><ul><li>In the limit <span>$\mathrm{LLR}\to 0^+$</span>, all cases must come from the null distribution. This will be the case if <span>$\hat{\alpha} \geq \alpha_0$</span>, and if this is not the case, we set <span>$\hat{\alpha} = \alpha_0$</span>.</li><li>In the limit <span>$\mathrm{LLR}\to \infty$</span>, all cases must come from the alternative distribution. This will be the case if <span>$\hat{\beta} \leq \beta_0$</span>, and if this is not the case, we set <span>$\hat{\beta} = \beta_0$</span>.</li></ul><p>The entire procedure is implemented in the <code>fit_mixdist_mom</code> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.fit_mixdist_mom" href="#BioFindr.fit_mixdist_mom"><code>BioFindr.fit_mixdist_mom</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mixdist_mom(llr,ns,ng=1,test=:corr)</code></pre><p>Fit a two-component mixture distribution of two LBeta distributions to a vector of log-likelihood ratios <code>llr</code> using a method-of-moments algorithm. The first component is the true null distribution for a given BioFindr <code>test</code> with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The second component is the alternative distribution, assumed to follow an <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a> distribution. The prior probability <code>pi0</code> of an observation belonging to the null component is fixed and determined by the <a href="#BioFindr.pi0est"><code>pi0est</code></a> function. Hence only the parameters of the alternative component need to be estimated.</p><p>The input variable <code>test</code> can take the values:</p><ul><li>:corr - <strong>correlation test</strong> (test 0)</li><li>:link - <strong>linkage test</strong> (test 1/2)</li><li>:med - <strong>mediation test</strong> (test 3)</li><li>:relev - <strong>relevance test</strong> (test 4)</li><li>:pleio - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to <code>:corr</code>, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p><p>See also <a href="#BioFindr.pi0est"><code>pi0est</code></a>, <a href="#BioFindr.fit_mom"><code>fit_mom</code></a>, <a href="../randomLLR/#BioFindr.LBeta"><code>LBeta</code></a>, <a href="../randomLLR/#BioFindr.nulldist"><code>nulldist</code></a>, <a href="../randomLLR/#BioFindr.nullpval"><code>nullpval</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L151-L167">source</a></section></article><h2 id="kde_postprobs"><a class="docs-heading-anchor" href="#kde_postprobs">Kernel density estimation of the mixture distribution</a><a id="kde_postprobs-1"></a><a class="docs-heading-anchor-permalink" href="#kde_postprobs" title="Permalink"></a></h2><p>The method of moments mixture distribution works very well in practice (by visual inspection of histograms and fitted distributions), but there is no theory to support it. Moreover the condition on the moments to produce valid distribution parameters (See <a href="#BioFindr.fit_mom"><code>fit_mom</code></a>) is sometimes violated (typically when <span>$\pi_0\approx 1$</span>). Hence <a href="https://github.com/tmichoel/BioFindr.jl">BioFindr.jl</a> also implements an alternative <a href="https://en.wikipedia.org/wiki/Kernel_density_estimation">kernel density estimation</a> method, comparable to the one implemented in the <a href="https://github.com/StoreyLab/qvalue">qvalue package for R</a>.</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.fit_mixdist_KDE" href="#BioFindr.fit_mixdist_KDE"><code>BioFindr.fit_mixdist_KDE</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mixdist_KDE(llr,ns,[ng,test])</code></pre><p>Return posterior probabilities for a vector of log-likelihood ratio values <code>llr</code> for a given BioFindr <code>test</code> with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p><p>See also <a href="#BioFindr.pi0est"><code>pi0est</code></a>, <a href="#BioFindr.fit_kde"><code>fit_kde</code></a>, <a href="../randomLLR/#BioFindr.nulldist"><code>nulldist</code></a>, <a href="../randomLLR/#BioFindr.nullpval"><code>nullpval</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L218-L232">source</a></section></article><p>To mitigate edge effects due to LLRs being restricted to the half-open interval <span>$[0,\infty)$</span>, the kernel density estimation of <span>$p(\mathrm{LLR})$</span> operates on transformed values:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.fit_kde" href="#BioFindr.fit_kde"><code>BioFindr.fit_kde</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_kde(llr)</code></pre><p>Fit a distribution function to a vector of log-likelihood ratio values <code>llr</code> using kernel density estimation. To avoid boundary effects in the KDE, the log-likelihoods (which take values in <span>$[0,\infty)$</span>) are first transformed to a vector</p><p><span>$z = \log \left( e^{2 \mathrm{LLR}} - 1 \right)$</span></p><p>which takes values in <span>$(-\infty,\infty)$</span>. KDE is applied to the transformed variables, and the probability density function (pdf) for <code>llr</code> is obtained from the pdf for <span>$z$</span> by the usual transformation rule for functions of random variables.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/bd383b79e1f19884d5be579c88ba546db1c766f2/src/posteriorprobs.jl#L261-L271">source</a></section></article><p>Nevertheless, the KDE of <span>$p(\mathrm{LLR})$</span> will still fluctuate wildly near <span>$\mathrm{LLR}\approx 0$</span>. Given a set of samples <span>$x_1,\dots,x_m$</span> with KDE estimated PDF values <span>$\tilde{p}(x_1), \dots, \tilde{p}(x_m)$</span>, <a href="#BioFindr.fit_mixdist_KDE"><code>fit_mixdist_KDE</code></a> calculates smoothened and monotonically increasing estimayes <span>$p(x_1),\dots, p(x_m)$</span> as</p><p class="math-container">\[p(x_i) = \max\left(0, \min_{j\geq i} \tilde{p}(x_j) \right)\]</p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-Chen2007"><a class="tag is-link" href="#citeref-Chen2007">Chen2007</a>Chen L, Emmert-Streib F, Storey J. <a href="https://doi.org/10.1186/gb-2007-8-10-r219">Harnessing naturally randomized transcription to infer regulatory relationships among genes</a>. Genome Biol 8, R219 (2007).</li><li class="footnote" id="footnote-Storey2003"><a class="tag is-link" href="#citeref-Storey2003">Storey2003</a>Storey JD, Tibshirani R. <a href="https://doi.org/10.1073/pnas.1530509100">Statistical significance for genomewide studies</a>. Proceedings of the National Academy of Sciences. 2003;100(16):9440–9445.</li><li class="footnote" id="footnote-Wang2017"><a class="tag is-link" href="#citeref-Wang2017">Wang2017</a>Wang L, Michoel T (2017) <a href="https://doi.org/10.1371/journal.pcbi.1005703">Efficient and accurate causal inference with hidden confounders from genome-transcriptome variation data</a>. PLoS Comput Biol 13(8): e1005703.</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../randomLLR/">« Null distributions of the LLRs</a><a class="docs-footer-nextpage" href="../testLLR/">Tests to evaluate »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Friday 26 April 2024 09:20">Friday 26 April 2024</span>. Using Julia version 1.10.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
