<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Likelihood ratio tests · Findr.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/Findr.jl/realLLR/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">Findr.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../inference/">General inference algorithm</a></li><li class="is-active"><a class="tocitem" href>Likelihood ratio tests</a><ul class="internal"><li><a class="tocitem" href="#Correlation-test"><span>Correlation test</span></a></li><li><a class="tocitem" href="#Primary-(linkage)-test"><span>Primary (linkage) test</span></a></li></ul></li><li><a class="tocitem" href="../randomLLR/">Null distributions of the LLRs</a></li><li><a class="tocitem" href="../posteriorprobs/">Bayesian inference of posterior probabilities</a></li><li><a class="tocitem" href="../testLLR/">Tests to evaluate</a></li><li><a class="tocitem" href="../listfunctions/">List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Likelihood ratio tests</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Likelihood ratio tests</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/Findr.jl/blob/main/docs/src/realLLR.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Likelihood-ratio-tests"><a class="docs-heading-anchor" href="#Likelihood-ratio-tests">Likelihood ratio tests</a><a id="Likelihood-ratio-tests-1"></a><a class="docs-heading-anchor-permalink" href="#Likelihood-ratio-tests" title="Permalink"></a></h1><p>Consider correlated genes <span>$A$</span>, <span>$B$</span>, and a third variable <span>$E$</span> upstream of <span>$A$</span> and <span>$B$</span>, such as a significant eQTL of <span>$A$</span>. The eQTLs can be obtained either <em>de novo</em> using eQTL identification tools such as matrix-eQTL [@Shabalin:2012] or kruX [@Qi:2014], or from published analyses. Throughout this article, we assume that <span>$E$</span> is a significant eQTL of <span>$A$</span>, whereas extension to other data types is straightforward. We use <span>$A_i$</span> and <span>$B_i$</span> for the expression levels of gene <span>$A$</span> and <span>$B$</span> respectively, which are assumed to have gone through the supernormalization, and optionally the genotypes of the best eQTL of <span>$A$</span> as <span>$E_i$</span>, where <span>$i=1,\dots,n$</span> across samples. Genotypes are assumed to have a total of <span>$n_a$</span> alleles, so <span>$E_i\in\{0,\dots,n_a\}$</span>. We define the null and alternative hypotheses for a total of six tests, as shown in <strong>Table <a href="#tab-tests">1</a>{reference-type=&quot;ref&quot; reference=&quot;tab-tests&quot;}</strong>. LLRs of every test are calculated separately as follows:</p><h2 id="Correlation-test"><a class="docs-heading-anchor" href="#Correlation-test">Correlation test</a><a id="Correlation-test-1"></a><a class="docs-heading-anchor-permalink" href="#Correlation-test" title="Permalink"></a></h2><p>Define the null hypothesis as <span>$A$</span> and <span>$B$</span> are independent, and the alternative hypothesis as they are correlated:</p><p class="math-container">\[{\mathcal H}_{\mathrm{null}}^{\mathrm{(0)}}=A\qquad B,\hspace{4em}{\mathcal H}_{\mathrm{alt}}^{\mathrm{(0)}}=A \to B.\]</p><p>The superscript <span>$(0)$</span> is the numbering of the test. Both hypotheses are modeled with gene expression levels following bivariate normal distributions, as</p><p class="math-container">\[\begin{pmatrix}
    A_i\\\\
    B_i
\end{pmatrix} \sim 
N\left(
    \begin{pmatrix}
        0\\\\
        0
    \end{pmatrix},
    \begin{pmatrix}
        \sigma_{A0}^2 &amp; \rho\,\sigma_{A0}\sigma_{B0}\\\\
        \rho\,\sigma_{A0}\sigma_{B0} &amp; \sigma_{B0}^2
    \end{pmatrix}
\right)\]</p><p>for <span>$i=1,\dots,n$</span>. The null hypothesis corresponds to <span>$\rho=0$</span>.</p><p>Maximum likelihood estimators (MLE) for the model parameters <span>$\rho$</span>, <span>$\sigma_{A0}$</span>, and <span>$\sigma_{B0}$</span> are</p><p class="math-container">\[\hat{\rho}=\frac{1}{n}\sum_{i=1}^n A_i B_i,\qquad\hat{\sigma}_{A0}=\hat{\sigma}_{B0}=1,\]</p><p>and the LLR is simply</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(0)}}=-\frac{n}{2}\log(1-\hat{\rho}^2).\]</p><p>In the absence of genotype information, we use nonzero correlation between <span>$A$</span> and <span>$B$</span> as the indicator for <span>$A\rightarrow B$</span> regulation, giving the posterior probability</p><p class="math-container">\[P(A \to B)=P(\mathcal{H}_{\mathrm{alt}}^{\mathrm{(0)}} \mid \mathrm{LLR}^{\mathrm{(0)}}).\]</p><p>false negative</p><h2 id="Primary-(linkage)-test"><a class="docs-heading-anchor" href="#Primary-(linkage)-test">Primary (linkage) test</a><a id="Primary-(linkage)-test-1"></a><a class="docs-heading-anchor-permalink" href="#Primary-(linkage)-test" title="Permalink"></a></h2><p>Verify that <span>$E$</span> regulates <span>$A$</span> from <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}\equiv E\rightarrow A$</span> and <span>${\mathcal H}_{\mathrm{null}}^{\mathrm{(1)}}\equiv E\hspace{1.6em}A$</span>. For <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}$</span>, we model <span>$E\rightarrow A$</span> as <span>$A$</span> follows a normal distribution whose mean is determined by <span>$E$</span> categorically, i.e.</p><p class="math-container">\[A_i\mid E_i\sim N(\mu_{E_i},\sigma_A^2).\label{eq-ba-hm-d1}\]</p><p>From the total likelihood <span>$p(A\mid E)=\prod_{i=1}^np(A_i\mid E_i)$</span>, we find MLEs for model parameters <span>$\mu_j,j=0,1,\dots,n_a$</span>, and <span>$\sigma_A$</span>, as</p><p class="math-container">\[\hat\mu_j=\frac{1}{n_j}\sum_{i=1}^nA_i\delta_{E_ij},\hspace{2em}\hat\sigma_A^2=1-\sum_{j=0}^{n_a}\frac{n_j}{n}\hat\mu_j^2,\nonumber\]</p><p>where <span>$n_j$</span> is the sample count by genotype category,</p><p class="math-container">\[n_j\equiv\sum_{i=1}^n\delta_{E_ij}.\nonumber\]</p><p>The Kronecker delta function is defined as <span>$\delta_{xy}=1$</span> for <span>$x=y$</span>, and 0 otherwise. When summing over all genotype values (<span>$j=0,\dots,n_a$</span>), we only pick those that exist (<span>$n_j&gt;0$</span>) throughout this article. Since the null hypothesis is simply that <span>$A_i$</span> is sampled from a genotype-independent normal distribution, with MLEs of mean zero and standard deviation one due to the supernormalization (<strong>Section <a href="#ssec-algorithm">[ssec-algorithm]</a>{reference-type=&quot;ref&quot;    reference=&quot;ssec-algorithm&quot;}</strong>), the LLR for test 1 becomes</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(1)}}=-\frac{n}{2}\ln\hat\sigma_A^2.\label{eq-ba-hm-llr1}\]</p><p>By favoring a large <span>$\mathrm{LLR}^{\mathrm{(1)}}$</span>, we select  <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}$</span> and verify that <span>$E$</span> regulates <span>$A$</span>, with</p><p class="math-container">\[P(E\rightarrow A)=P({\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}\mid\mathrm{LLR}^{\mathrm{(1)}}).\nonumber\]</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../inference/">« General inference algorithm</a><a class="docs-footer-nextpage" href="../randomLLR/">Null distributions of the LLRs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 30 May 2023 20:06">Tuesday 30 May 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
