<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Likelihood ratio tests · BioFindr.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/BioFindr.jl/realLLR/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioFindr.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li><a class="tocitem" href="../inference/">General inference algorithm</a></li><li class="is-active"><a class="tocitem" href>Likelihood ratio tests</a><ul class="internal"><li><a class="tocitem" href="#corr_test_realLLR"><span>Correlation test</span></a></li><li><a class="tocitem" href="#prim_test_realLLR"><span>Primary linkage test</span></a></li><li><a class="tocitem" href="#Secondary-linkage-test"><span>Secondary linkage test</span></a></li><li><a class="tocitem" href="#Conditional-independence-test"><span>Conditional independence test</span></a></li><li><a class="tocitem" href="#Relevance-test"><span>Relevance test</span></a></li><li><a class="tocitem" href="#Controlled-or-pleiotropy-test"><span>Controlled or pleiotropy test</span></a></li><li><a class="tocitem" href="#Implementation"><span>Implementation</span></a></li></ul></li><li><a class="tocitem" href="../randomLLR/">Null distributions of the LLRs</a></li><li><a class="tocitem" href="../posteriorprobs/">Bayesian inference of posterior probabilities</a></li><li><a class="tocitem" href="../testLLR/">Tests to evaluate</a></li><li><a class="tocitem" href="../utils/">Utilities</a></li><li><a class="tocitem" href="../listfunctions/">List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Likelihood ratio tests</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Likelihood ratio tests</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/BioFindr.jl/blob/main/docs/src/realLLR.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Likelihood-ratio-tests"><a class="docs-heading-anchor" href="#Likelihood-ratio-tests">Likelihood ratio tests</a><a id="Likelihood-ratio-tests-1"></a><a class="docs-heading-anchor-permalink" href="#Likelihood-ratio-tests" title="Permalink"></a></h1><p>Consider correlated genes <span>$A$</span>, <span>$B$</span>, and a third variable <span>$E$</span> upstream of <span>$A$</span> and <span>$B$</span>, such as a significant eQTL of <span>$A$</span>. The eQTLs can be obtained either <em>de novo</em> using eQTL identification tools such as matrix-eQTL <sup class="footnote-reference"><a id="citeref-Shabalin2012" href="#footnote-Shabalin2012">[Shabalin2012]</a></sup> or kruX <sup class="footnote-reference"><a id="citeref-Qi2014" href="#footnote-Qi2014">[Qi2014]</a></sup>, or from published analyses. Throughout this documentation, we assume that <span>$E$</span> is a significant eQTL of <span>$A$</span>, whereas extension to other data types is straightforward. We use <span>$A_i$</span> and <span>$B_i$</span> for the expression levels of gene <span>$A$</span> and <span>$B$</span> respectively, which are assumed to have gone through the supernormalization, and optionally the genotypes of the best eQTL of <span>$A$</span> as <span>$E_i$</span>, where <span>$i=1,\dots,n$</span> across samples. Genotypes are assumed to have a total of <span>$n_a$</span> categories, so <span>$E_i\in\{0,\dots,n_a\}$</span>. We define the null and alternative hypotheses for a total of six tests, as shown in the table below. </p><div class="admonition is-info"><header class="admonition-header">Six likelihood ratio tests are performed to test the regulation A → B, numbered, named, and defined as shown.</header><div class="admonition-body"><p><img src="../pcbi.1005703.g001.png" alt="Six likelihood ratio tests"/></p><p>Copyright 2017 Wang, Michoel. <a href="https://doi.org/10.1371/journal.pcbi.1005703.g001">DOI:10.1371/journal.pcbi.1005703.g001</a>. Reused under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution License</a></p></div></div><p>LLRs of every test are calculated separately as follows:</p><h2 id="corr_test_realLLR"><a class="docs-heading-anchor" href="#corr_test_realLLR">Correlation test</a><a id="corr_test_realLLR-1"></a><a class="docs-heading-anchor-permalink" href="#corr_test_realLLR" title="Permalink"></a></h2><p>Define the null hypothesis as <span>$A$</span> and <span>$B$</span> are independent, and the alternative hypothesis as they are correlated:</p><p class="math-container">\[{\mathcal H}_{\mathrm{null}}^{\mathrm{(0)}}=A\qquad B,\hspace{4em}{\mathcal H}_{\mathrm{alt}}^{\mathrm{(0)}}=A - B.\]</p><p>The superscript <span>$(0)$</span> is the numbering of the test. Both hypotheses are modeled with gene expression levels following bivariate normal distributions, as</p><p class="math-container">\[\begin{pmatrix}
    A_i\\\\
    B_i
\end{pmatrix} \sim 
N\left(
    \begin{pmatrix}
        0\\\\
        0
    \end{pmatrix},
    \begin{pmatrix}
        \sigma_{A0}^2 &amp; \rho\,\sigma_{A0}\sigma_{B0}\\\\
        \rho\,\sigma_{A0}\sigma_{B0} &amp; \sigma_{B0}^2
    \end{pmatrix}
\right)\]</p><p>for <span>$i=1,\dots,n$</span>. The null hypothesis corresponds to <span>$\rho=0$</span>.</p><p>Maximum likelihood estimators (MLE) for the model parameters <span>$\rho$</span>, <span>$\sigma_{A0}$</span>, and <span>$\sigma_{B0}$</span> are</p><p class="math-container">\[\hat{\rho}=\frac{1}{n}\sum_{i=1}^n A_i B_i,\qquad\hat{\sigma}_{A0}=\hat{\sigma}_{B0}=1,\]</p><p>and the LLR is simply</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(0)}}=-\frac{n}{2}\log(1-\hat{\rho}^2).\]</p><p>In the absence of genotype information, we use nonzero correlation between <span>$A$</span> and <span>$B$</span> as the indicator for <span>$A\rightarrow B$</span> regulation, giving the posterior probability</p><p class="math-container">\[P(A - B)=P(\mathcal{H}_{\mathrm{alt}}^{\mathrm{(0)}} \mid \mathrm{LLR}^{\mathrm{(0)}}).\]</p><p>The LLR for the correlation test of a specific gene <span>$A$</span> against all other genes <span>$B$</span> is implemented as a method of the <code>realLLR_col</code> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat" href="#BioFindr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat"><code>BioFindr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Matrix{T},Ycol::Vector{T}) where T&lt;:AbstractFloat</code></pre><p>Compute the log-likelihood ratios for BioFindr test 0 (<strong>correlation test</strong>) for a given column vector <code>Ycol</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L1-L9">source</a></section></article><h2 id="prim_test_realLLR"><a class="docs-heading-anchor" href="#prim_test_realLLR">Primary linkage test</a><a id="prim_test_realLLR-1"></a><a class="docs-heading-anchor-permalink" href="#prim_test_realLLR" title="Permalink"></a></h2><p>Verify that <span>$E$</span> regulates <span>$A$</span> from <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}\equiv E\rightarrow A$</span> and <span>${\mathcal H}_{\mathrm{null}}^{\mathrm{(1)}}\equiv E\qquad A$</span>. For <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}$</span>, we model <span>$E\rightarrow A$</span> as <span>$A$</span> follows a normal distribution whose mean is determined by <span>$E$</span> categorically, i.e.</p><p class="math-container">\[A_i\mid E_i\sim N(\mu_{E_i},\sigma_A^2).\]</p><p>From the total likelihood <span>$p(A\mid E)=\prod_{i=1}^np(A_i\mid E_i)$</span>, we find MLEs for model parameters <span>$\mu_j,j=0,1,\dots,n_a$</span> and <span>$\sigma_A$</span>, as</p><p class="math-container">\[\hat{\mu}_j=\frac{1}{n_j}\sum_{i=1}^n A_i \delta_{E_i j},\quad \hat{\sigma}_A^2=1-\sum_{j=0}^{n_a}\frac{n_j}{n} \hat{\mu}_j^2,\]</p><p>where <span>$n_j$</span> is the sample count by genotype category,</p><p class="math-container">\[n_j \equiv \sum_{i=1}^n \delta_{E_i j}.\]</p><p>The Kronecker delta function is defined as <span>$\delta_{xy}=1$</span> for <span>$x=y$</span>, and 0 otherwise. When summing over all genotype values (<span>$j=0,\dots,n_a$</span>), we only pick those that exist (<span>$n_j&gt;0$</span>) throughout this article. Since the null hypothesis is simply that <span>$A_i$</span> is sampled from a genotype-independent normal distribution, with MLEs of mean zero and standard deviation one due to the supernormalization (see <a href="../inference/#General-inference-algorithm">General inference algorithm</a>), the LLR for test 1 becomes</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(1)}}=-\frac{n}{2}\ln\hat{\sigma}_A^2.\]</p><p>By favoring a large <span>$\mathrm{LLR}^{\mathrm{(1)}}$</span>, we select  <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}$</span> and verify that <span>$E$</span> regulates <span>$A$</span>, with</p><p class="math-container">\[P(E\rightarrow A)=P({\mathcal H}_{\mathrm{alt}}^{\mathrm{(1)}}\mid\mathrm{LLR}^{\mathrm{(1)}}).\]</p><p>The LLR for the primary linkage test can be obtained by selecting the appropriate element from the results of the secondary linkage test.</p><h2 id="Secondary-linkage-test"><a class="docs-heading-anchor" href="#Secondary-linkage-test">Secondary linkage test</a><a id="Secondary-linkage-test-1"></a><a class="docs-heading-anchor-permalink" href="#Secondary-linkage-test" title="Permalink"></a></h2><p>The secondary test is identical with the primary test, except it verifies that <span>$E$</span> regulates <span>$B$</span>. Hence repeat the primary test on <span>$E$</span> and <span>$B$</span> and obtain the MLEs:</p><p class="math-container">\[\hat{\nu}_j=\frac{1}{n_j}\sum_{i=1}^n B_i \delta_{E_i j},\quad \hat{\sigma}_B^2=1-\sum_{j=0}^{n_a}\frac{n_j}{n}\hat{\nu}_j^2,\]</p><p>and the LLR as</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(2)}}=-\frac{n}{2}\ln\hat{\sigma}_B^2.\]</p><p><span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(2)}}$</span> is chosen to verify that <span>$E$</span> regulates <span>$B$</span>.</p><p>In <a href="../testLLR/#Association-analysis">Association analysis</a>, the linkage test is used standalone, and then its LLR for testing a specific grouping vector <span>$E$</span> against all genes <span>$B$</span> is implemented as a method of the <code>realLLR_col</code> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>BioFindr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Array{T},E::Vector{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the log-likelihood ratios for BioFindr test 2 (<strong>Linkage test</strong>)  for a given categorical vector <code>E</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.llrstats_col"><code>llrstats_col</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L55-L63">source</a></section></article><p>In <a href="../testLLR/#Causal-inference">Causal inference</a>, the LLR for the linkage test is computed together with the other tests for efficiency (see below).</p><h2 id="Conditional-independence-test"><a class="docs-heading-anchor" href="#Conditional-independence-test">Conditional independence test</a><a id="Conditional-independence-test-1"></a><a class="docs-heading-anchor-permalink" href="#Conditional-independence-test" title="Permalink"></a></h2><p>Verify that <span>$E$</span> and <span>$B$</span> are independent when conditioning on <span>$A$</span>. This can be achieved by comparing </p><p class="math-container">\[{\mathcal H}_{\mathrm{alt}}^{\mathrm{(3)}}\equiv B\leftarrow E\rightarrow A\wedge(A\mathrm{\ correlates\ with\ }B)\]</p><p>against </p><p class="math-container">\[{\mathcal H}_{\mathrm{null}}^{\mathrm{(3)}}\equiv E\rightarrow A\rightarrow B.\]</p><p>LLRs close to zero then prefer <span>${\mathcal H}_{\mathrm{null}}^{\mathrm{(3)}}$</span>, and ensure that <span>$E$</span> regulates <span>$B$</span> only through <span>$A$</span>:</p><p class="math-container">\[P(E\perp B\mid A)=P({\mathcal H}_{\mathrm{null}}^{\mathrm{(3)}}\mid\mathrm{LLR}^{\mathrm{(3)}}).\]</p><p>For <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(3)}}$</span>, the bivariate normal distribution dependent on <span>$E$</span> can be represented as</p><p class="math-container">\[\begin{pmatrix}
    A_i\\\\
    B_i
\end{pmatrix}
\mid E_i 
\sim N\left(
    \begin{pmatrix}
        \mu_{E_i}\\\\
        \nu_{E_i}
    \end{pmatrix},
    \begin{pmatrix}
        \sigma_A^2 &amp; \rho\sigma_A\sigma_B \\\\
        \rho\sigma_A\sigma_B &amp; \sigma_B^2
    \end{pmatrix}
    \right).\]</p><p>For <span>${\mathcal H}_{\mathrm{null}}^{\mathrm{(3)}}$</span>, the distributions follow <span>$A_i\mid E_i$</span> as in the <a href="#prim_test_realLLR">Primary linkage test</a>, as well as</p><p class="math-container">\[B_i\mid A_i\sim N(\rho A_i,\sigma_B^2).\]</p><p>Substituting parameters <span>$\mu_j,\nu_j,\sigma_A,\sigma_B,\rho$</span> of <span>${\mathcal H}_{\mathrm{alt}}^{\mathrm{(3)}}$</span> and <span>$\mu_j,\rho,\sigma_A,\sigma_B$</span> of <span>${\mathcal H}_{\mathrm{null}}^{\mathrm{(3)}}$</span> with their MLEs, we obtain the LLR: </p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(3)}} =-\frac{n}{2}\ln\left(\hat{\sigma}_A^2\hat{\sigma}_B^2-(\hat{\rho}+\sigma_{AB}-1)^2\right) + \frac{n}{2}\ln\hat{\sigma}_A^2 + \frac{n}{2}\ln(1-\hat{\rho}^2),\]</p><p>where</p><p class="math-container">\[\sigma_{AB} \equiv 1-\sum_{j=0}^{n_a}\frac{n_j}{n}\hat{\mu}_j\hat{\nu}_j,\]</p><p>and <span>$\hat\rho=\frac{1}{n}\sum_{i=1}^n A_iB_i$</span> is the same as in the <a href="#corr_test_realLLR">Correlation test</a>.</p><h2 id="Relevance-test"><a class="docs-heading-anchor" href="#Relevance-test">Relevance test</a><a id="Relevance-test-1"></a><a class="docs-heading-anchor-permalink" href="#Relevance-test" title="Permalink"></a></h2><p>Since the indirect regulation <span>$E\rightarrow B$</span> tends to be weaker than any of its direct regulation components (<span>$E\rightarrow A$</span> or <span>$A\rightarrow B$</span>), we propose to test <span>$E\rightarrow A\rightarrow B$</span> with indirect regulation <span>$E\rightarrow B$</span> as well as the direct regulation <span>$A\rightarrow B$</span> for stronger distinguishing power on weak regulations. We define</p><p class="math-container">\[{\mathcal H}_{\mathrm{alt}}^{\mathrm{(4)}}\equiv E\rightarrow A\wedge E\rightarrow B\leftarrow A\]</p><p>and</p><p class="math-container">\[{\mathcal H}_{\mathrm{null}}^{\mathrm{(4)}}\equiv E\rightarrow A\qquad B\]</p><p>This simply verifies that <span>$B$</span> is not independent from both <span>$A$</span> and <span>$E$</span> simultaneously. In the alternative hypothesis, <span>$B$</span> is regulated by <span>$E$</span> and <span>$A$</span>, which is modeled as a normal distribution whose mean is additively determined by <span>$E$</span> categorically and <span>$A$</span> linearly, i.e.</p><p class="math-container">\[B_i\mid E_i,A_i\sim N(\nu_{E_i}+\rho A_i,\sigma_B^2).\]</p><p>We can hence solve its LLR as</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(4)}}=-\frac{n}{2}\ln\left(\hat{\sigma}_A^2\hat{\sigma}_B^2-(\hat{\rho}+\sigma_{AB}-1)^2\right)+\frac{n}{2}\ln\hat{\sigma}_A^2,\]</p><p>with all MLEs as defined before.</p><h2 id="Controlled-or-pleiotropy-test"><a class="docs-heading-anchor" href="#Controlled-or-pleiotropy-test">Controlled or pleiotropy test</a><a id="Controlled-or-pleiotropy-test-1"></a><a class="docs-heading-anchor-permalink" href="#Controlled-or-pleiotropy-test" title="Permalink"></a></h2><p>Based on the positives of the secondary test, we can further distinguish the alternative hypothesis </p><p class="math-container">\[{\mathcal H}_{\mathrm{alt}}^{\mathrm{(5)}}\equiv B\leftarrow E\rightarrow A\wedge A\rightarrow B\]</p><p>from the null</p><p class="math-container">\[{\mathcal H}_{\mathrm{null}}^{\mathrm{(5)}}\equiv B\leftarrow E\rightarrow A\]</p><p>to verify that <span>$E$</span> does not regulate <span>$A$</span> and <span>$B$</span> independently. Its LLR can be solved as</p><p class="math-container">\[\mathrm{LLR}^{\mathrm{(5)}}=-\frac{n}{2}\ln\left(\hat{\sigma}_A^2\hat{\sigma}_B^2-(\hat{\rho}+\sigma_{AB}-1)^2\right)+\frac{n}{2}\ln\hat{\sigma}_A^2\hat\sigma_B^2.\]</p><p>with all MLEs as defined before. Note that this test was called the <strong>controlled</strong> test in the original paper (see figure above), but we now prefer to refer to it as the <strong>pleiotropy</strong> test.</p><h2 id="Implementation"><a class="docs-heading-anchor" href="#Implementation">Implementation</a><a id="Implementation-1"></a><a class="docs-heading-anchor-permalink" href="#Implementation" title="Permalink"></a></h2><p>The LLRs for the secondary linkage, conditional independence, relevance, and pleiotrropy tests of a specific gene <span>$A$</span> against all other genes <span>$B$</span> can all be computed from the same sufficient statistics <span>$\hat{\rho}, \hat{\sigma}_A^2, \hat{\sigma}_B^2, \sigma_{AB}$</span>. Moreover note the relations</p><p class="math-container">\[\begin{aligned}
\mathrm{LLR}^{\mathrm{(3)}} &amp;= \mathrm{LLR}^{\mathrm{(4)}} + \frac{n}{2}\ln (1-\hat{\rho}^2)\\
\mathrm{LLR}^{\mathrm{(5)}} &amp;= \mathrm{LLR}^{\mathrm{(4)}} + \frac{n}{2}\ln \hat{\sigma}_B^2 = \mathrm{LLR}^{\mathrm{(4)}} - \mathrm{LLR}^{\mathrm{(2)}}
\end{aligned}\]</p><p>Hence for efficiency these LLRs are computed jointly by a method of the <code>realLLR_col</code> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#BioFindr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>BioFindr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the log-likelihood ratios for the BioFindr causal tests for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of matrix <code>Y</code> : </p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.llrstats_col"><code>llrstats_col</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L16-L29">source</a></section></article><p>The MLEs of the various model parameters are computed in the <code>llrstats_col</code> and <code>groupmeans</code> functions:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.llrstats_col" href="#BioFindr.llrstats_col"><code>BioFindr.llrstats_col</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">llrstats_col(Y,Ycol,E)</code></pre><p>Compute the sufficient statistics for the log-likelihood ratios for BioFindr tests 2-5  for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of <code>Y</code>.</p><p><code>Ycol</code> and <code>Y</code> are assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>The sufficient statistics are:</p><ul><li>the covariance <span>$\hat{\rho}$</span> between <code>Ycol</code> and all columns of <code>Y</code>, returned in the first output argument,</li><li>the weighted average variances <span>$\hat{\sigma}_B^2$</span> of each column of matrix <code>Y</code> over the groups (unique values) in <code>E</code>, returned as the first column of the second output argument,</li><li>the weighted average covariances <span>$\sigma_{AB}$</span> between <code>Ycol</code> and all  columns of <code>Y</code> over the groups (unique values) of <code>E</code>, returned as the second column of the second output argument,</li><li>the weighted average variance <span>$\hat{\sigma}_A^2$</span> of <code>Ycol</code> over the groups (unique values) in <code>E</code>, returned as the third output argument.</li></ul><p>See also <a href="#BioFindr.groupmeans"><code>groupmeans</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L72-L87">source</a></section><section><div><pre><code class="nohighlight hljs">llrstats_col(Y,E)</code></pre><p>Compute the sufficient statistics for the log-likelihood ratios for BioFindr tests 2  for a given categorical vector <code>E</code> against all columns of <code>Y</code>.</p><p><code>Y</code> is assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>The sufficient statistics are the weighted average variances <span>$\hat{\sigma}_B^2$</span> of each column of matrix <code>Y</code> over the groups (unique values) in <code>E</code>.</p><p>See also <a href="#BioFindr.groupmeans"><code>groupmeans</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L103-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioFindr.groupmeans" href="#BioFindr.groupmeans"><code>BioFindr.groupmeans</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">groupmeans(Y,Ycol,E)</code></pre><p>Compute the size <span>$n_j$</span> for each of the groups (unique values) in categorical vector <code>E</code> (first output argument), and the means <span>$\hat{\nu_j}$</span> of each column of matrix <code>Y</code> (second output argument) and <span>$\hat{\mu_j}$</span> of <code>Ycol</code> (third output argument) for each of the groups in <code>E</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L123-L127">source</a></section><section><div><pre><code class="nohighlight hljs">groupmeans(Y,E)</code></pre><p>Compute the size <span>$n_j$</span> for each of the groups (unique values) in categorical vector <code>E</code> (first output argument), and the means <span>$\hat{\nu_j}$</span> of each column of matrix <code>Y</code> (second output argument) for each of the groups in <code>E</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L142-L146">source</a></section></article><p>A summary of all methods of the <code>realLLR_col</code> function:</p><article class="docstring"><header><a class="docstring-binding" id="BioFindr.realLLR_col" href="#BioFindr.realLLR_col"><code>BioFindr.realLLR_col</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Matrix{T},Ycol::Vector{T}) where T&lt;:AbstractFloat</code></pre><p>Compute the log-likelihood ratios for BioFindr test 0 (<strong>correlation test</strong>) for a given column vector <code>Ycol</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L1-L9">source</a></section><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Matrix{T},Ycol::Vector{T},E::Vector{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the log-likelihood ratios for the BioFindr causal tests for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of matrix <code>Y</code> : </p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.llrstats_col"><code>llrstats_col</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L16-L29">source</a></section><section><div><pre><code class="language-julia hljs">realLLR_col(Y::Array{T},E::Vector{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute the log-likelihood ratios for BioFindr test 2 (<strong>Linkage test</strong>)  for a given categorical vector <code>E</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>See also <a href="../inference/#BioFindr.supernormalize"><code>supernormalize</code></a>, <a href="#BioFindr.llrstats_col"><code>llrstats_col</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/BioFindr.jl/blob/336d83bd8bf11c9cc2c0088a878ea07ad38e1a4e/src/realLLR.jl#L55-L63">source</a></section></article><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-Shabalin2012"><a class="tag is-link" href="#citeref-Shabalin2012">Shabalin2012</a>Shabalin AA. <a href="https://doi.org/10.1093/bioinformatics/bts163">Matrix eQTL: ultra fast eQTL analysis via large matrix operations</a>, Bioinformatics, Volume 28, Issue 10, May 2012, Pages 1353–1358, </li><li class="footnote" id="footnote-Qi2014"><a class="tag is-link" href="#citeref-Qi2014">Qi2014</a>Qi J, Foroughi Asl H, Bjorkegren J, Michoel T. <a href="https://doi.org/10.1186/1471-2105-15-11">kruX: matrix-based non-parametric eQTL discovery</a>. BMC Bioinformatics 15, 11 (2014).</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../inference/">« General inference algorithm</a><a class="docs-footer-nextpage" href="../randomLLR/">Null distributions of the LLRs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 6 February 2024 09:29">Tuesday 6 February 2024</span>. Using Julia version 1.10.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
