<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>General inference algorithm · Findr.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/Findr.jl/inference/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">Findr.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li class="is-active"><a class="tocitem" href>General inference algorithm</a><ul class="internal"><li><a class="tocitem" href="#Coexpression-analysis"><span>Coexpression analysis</span></a></li><li><a class="tocitem" href="#Differential-expression-analysis"><span>Differential expression analysis</span></a></li><li><a class="tocitem" href="#Causal-inference"><span>Causal inference</span></a></li><li><a class="tocitem" href="#Bipartite-causal-inference"><span>Bipartite causal inference</span></a></li></ul></li><li><a class="tocitem" href="../realLLR/">Likelihood ratio tests</a></li><li><a class="tocitem" href="../randomLLR/">Null distributions of the LLRs</a></li><li><a class="tocitem" href="../posteriorprobs/">Bayesian inference of posterior probabilities</a></li><li><a class="tocitem" href="../testLLR/">Tests to evaluate</a></li><li><a class="tocitem" href="../bayesiannets/">Bayesian networks</a></li><li><a class="tocitem" href="../utils/">Utilities</a></li><li><a class="tocitem" href="../listfunctions/">List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>General inference algorithm</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>General inference algorithm</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/Findr.jl/blob/main/docs/src/inference.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="General-inference-algorithm"><a class="docs-heading-anchor" href="#General-inference-algorithm">General inference algorithm</a><a id="General-inference-algorithm-1"></a><a class="docs-heading-anchor-permalink" href="#General-inference-algorithm" title="Permalink"></a></h1><p>Consider a set of observations sampled from a mixture distribution of a null and an alternative hypothesis. For instance in gene regulation, every observation can correspond to expression levels of a pair of genes wich are sampled from a bivariate normal distribution with zero (null hypothesis) or non-zero (alternative hypothesis) correlation coefficient. In Findr, we predict the probability that any sample follows the alternative hypothesis with the following algorithm (based on and modified from <sup class="footnote-reference"><a id="citeref-Chen2007" href="#footnote-Chen2007">[Chen2007]</a></sup>):</p><ol><li>For robustness against outliers, we convert every continuous variable into standard normally distributed <span>$N(0,1)$</span> values using a rank-based inverse normal transformation across all samples. We name this step as <em>supernormalization</em>.</li><li>We propose a null and an alternative hypothesis for every likelihood ratio test (LRT) of interest where, by definition, the null hypothesis space is a subset of the alternative hypothesis. Model parameters are replaced with their maximum likelihood estimators (MLEs) to obtain the log likelihood ratio (LLR) between the alternative and null hypotheses.</li><li>We derive the analytical expression for the probablity density function (PDF) of the LLR when samples follow the null hypothesis.</li><li>We convert LLRs into posterior probabilities of the hypothesis of interest with the empirical estimation of local FDR.</li></ol><p>While primarily developed for causal inference, several common tasks in genome-wide studies can be implemented using this algorithm and are supported in Findr. </p><h2 id="Coexpression-analysis"><a class="docs-heading-anchor" href="#Coexpression-analysis">Coexpression analysis</a><a id="Coexpression-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Coexpression-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat" href="#Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T}) where T&lt;:AbstractFloat</code></pre><p>Compute posterior probabilities for nonzero pairwise correlations between columns of input matrix <code>X</code>. The probabilities are directed (asymmetric) in the sense that they are estimated from a column-specific background distribution. </p><p>See also <a href="#Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame"><code>findr(dX::DataFrame)</code></a>-</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L42-L48">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame" href="#Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(dX::T) where T&lt;:AbstractDataFrame</code></pre><p>Wrapper for <code>findr(Matrix(dX))</code> when the input <code>dX</code> is in the form of a DataFrame. The output is then also wrapped in a DataFrame with the same column names. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L62-L66">source</a></section></article><h2 id="Differential-expression-analysis"><a class="docs-heading-anchor" href="#Differential-expression-analysis">Differential expression analysis</a><a id="Differential-expression-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Differential-expression-analysis" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T},G::Array{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero differential expression of colunns of input matrix <code>X</code> across groups defined by one or more categorical variables (columns of <code>G</code>).</p><p>Return a matrix of size ncols(X) x ncols(G)</p><p>See also <a href="#Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat"><code>findr(dX::DataFrame,dG::DataFrame)</code></a>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>G</code> is currently assumed to be an array (vector or matrix) of integers. CategoricalArrays will be supported in the future.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L73-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{T}, Tuple{T, T}} where T&lt;:DataFrames.AbstractDataFrame" href="#Findr.findr-Union{Tuple{T}, Tuple{T, T}} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(dX::T, dG::T) where T&lt;:AbstractDataFrame</code></pre><p>Wrapper for <code>findr(Matrix(dX), Matrix(dG))</code> when the inputs <code>dX</code> and <code>dG</code> are in the form of a DataFrame. The output is then also wrapped in a DataFrame with the same column names. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L98-L102">source</a></section></article><h2 id="Causal-inference"><a class="docs-heading-anchor" href="#Causal-inference">Causal inference</a><a id="Causal-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Causal-inference" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{R}}} where {T&lt;:AbstractFloat, S&lt;:Integer, R&lt;:Integer}" href="#Findr.findr-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{R}}} where {T&lt;:AbstractFloat, S&lt;:Integer, R&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T},G::Matrix{S},pairGX::Matrix{S}; combination=&quot;none&quot;) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero causal relations between columns of input matrix <code>X</code>. The probabilities are estimated for a subset of columns of <code>X</code> that have a (discrete) instrumental variable in input matrix <code>G</code>. The matching between columns of <code>X</code> and columns of <code>G</code> is given by <code>pairGX</code>, a two-column array where the first column corresponds to a column index in <code>G</code> and the second to a column index in <code>X</code>.</p><p>Posterior probabilities are computed for the following tests</p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p>which can be combined into the mediation test (<span>$P_2 P_3$</span>; <code>combination=&quot;mediation&quot;</code>), the instrumental variable or non-independence test (<span>$P_2 P_5$</span>; <code>combination=&quot;IV&quot;</code>), or Findr&#39;s original combination (<span>$\frac{1}{2}(P_2 P_5 + P_4)$</span>; <code>combination=&quot;orig&quot;</code>). By default, individual probability matrices for all tests are returned (<code>combination=&quot;none&quot;</code>).</p><p>All return matrices have size ncols(X) x ncols(G).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>G</code> is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L110-L128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{T}, Tuple{T, T, T}} where T&lt;:DataFrames.AbstractDataFrame" href="#Findr.findr-Union{Tuple{T}, Tuple{T, T, T}} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(dX::T, dG::T, dE::T, colX=2, colG=1, combination=&quot;IV&quot;) where T&lt;:AbstractDataFrame</code></pre><p>Wrapper for <code>findr(X, G, pairGX)</code> when the inputs are in the form of a DataFrame. The output is then also wrapped in a DataFrame with the same column names. When DataFrames are used, only a combined posterior probabilities can be returned (<code>combination=&quot;IV&quot;</code> (default), <code>&quot;mediation&quot;</code>, or <code>&quot;orig&quot;</code>).</p><p>The input dataframes are:</p><ul><li><code>dX</code> - DataFrame with expression data, columns are genes</li><li><code>dG</code> - DataFrame with genotype data, columns are variants (SNPs)</li><li><code>dE</code> - DataFrame with eQTL results, must contains columns with gene and SNP IDs that can be mapped to column names in <code>dX</code> and <code>dG</code>, respectively</li><li><code>colG</code> - name or number of variant ID column in <code>dE</code>, default 1</li><li><code>colX</code> - name or number of gene ID column in <code>dE</code>, default 2</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L147-L159">source</a></section></article><h2 id="Bipartite-causal-inference"><a class="docs-heading-anchor" href="#Bipartite-causal-inference">Bipartite causal inference</a><a id="Bipartite-causal-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Bipartite-causal-inference" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}, Matrix{R}}} where {T&lt;:AbstractFloat, S&lt;:Integer, R&lt;:Integer}" href="#Findr.findr-Union{Tuple{R}, Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}, Matrix{R}}} where {T&lt;:AbstractFloat, S&lt;:Integer, R&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X1::Matrix{T},X2::Array{T},G::Array{S},pairGX::Matrix{R}; combination=&quot;none&quot;)  where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero causal relations from columns of input matrix <code>X2</code> to columns of input matrix <code>X1</code>. The probabilities are estimated for a subset of columns of <code>X2</code> that have a (discrete) instrumental variable in input matrix <code>G</code>. The matching between columns of <code>X2</code> and columns of <code>G</code> is given by <code>pairGX</code>, a two-column array where the first column corresponds to a column index in <code>G</code> and the second to a column index in <code>X2</code>.</p><p>Posterior probabilities are computed for the following tests</p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p>which can be combined into the mediation test (<span>$P_2 P_3$</span>; <code>combination=&quot;mediation&quot;</code>), the instrumental variable or non-independence test (<span>$P_2 P_5$</span>; <code>combination=&quot;IV&quot;</code>), or Findr&#39;s original combination (<span>$\frac{1}{2}(P_2 P_5 + P_4)$</span>; <code>combination=&quot;orig&quot;</code>). By default, individual probability matrices for all tests are returned (<code>combination=&quot;none&quot;</code>).</p><p>All return matrices have size ncols(X1) x ncols(X2).</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>G</code> is currently assumed to be an array (vector or matrix) of integers. I intend to use CategoricalArrays in the future.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L178-L196">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{T}, NTuple{4, T}} where T&lt;:DataFrames.AbstractDataFrame" href="#Findr.findr-Union{Tuple{T}, NTuple{4, T}} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(dX1::T, dX2::T, dG::T, dE::T; colG=1, colX=2, combination=&quot;IV&quot;) where T&lt;:AbstractDataFrame</code></pre><p>Wrapper for <code>findr(Matrix(dX1), Matrix(dX2), Matrix(dG))</code> when the inputs <code>dX1</code>, <code>dX2</code>, and <code>dG</code> are in the form of a DataFrame. The output is then also wrapped in a DataFrame with the same column names. When DataFrames are used, only a combined posterior probabilities can be returned (<code>combination=&quot;IV&quot;</code> (default), <code>&quot;mediation&quot;</code>, or <code>&quot;orig&quot;</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/81399405bea68d68c4a2923c9b0f69decf7ff27d/src/Findr.jl#L217-L221">source</a></section></article><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-Chen2007"><a class="tag is-link" href="#citeref-Chen2007">Chen2007</a>Chen L, Emmert-Streib F, Storey J. <a href="https://doi.org/10.1186/gb-2007-8-10-r219">Harnessing naturally randomized transcription to infer regulatory relationships among genes</a>. Genome Biol 8, R219 (2007).</li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Introduction</a><a class="docs-footer-nextpage" href="../realLLR/">Likelihood ratio tests »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Monday 5 June 2023 13:14">Monday 5 June 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
