<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>List of functions · Findr.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://tmichoel.github.io/Findr.jl/listfunctions/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">Findr.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../inference/">General inference algorithm</a></li><li><a class="tocitem" href="../realLLR/">Likelihood ratio tests</a></li><li><a class="tocitem" href="../randomLLR/">Null distributions of the LLRs</a></li><li><a class="tocitem" href="../posteriorprobs/">Bayesian inference of posterior probabilities</a></li><li><a class="tocitem" href="../testLLR/">Tests to evaluate</a></li><li class="is-active"><a class="tocitem" href>List of functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>List of functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>List of functions</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tmichoel/Findr.jl/blob/main/docs/src/listfunctions.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="List-of-package-functions"><a class="docs-heading-anchor" href="#List-of-package-functions">List of package functions</a><a id="List-of-package-functions-1"></a><a class="docs-heading-anchor-permalink" href="#List-of-package-functions" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/tmichoel/Findr.jl">Findr</a>.</p><ul><li><a href="#Distributions.ccdf-Tuple{Findr.LBeta, Real}"><code>Distributions.ccdf</code></a></li><li><a href="#Distributions.cdf-Tuple{Findr.LBeta, Real}"><code>Distributions.cdf</code></a></li><li><a href="#Distributions.fit_mle-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real"><code>Distributions.fit_mle</code></a></li><li><a href="#Distributions.logccdf-Tuple{Findr.LBeta, Real}"><code>Distributions.logccdf</code></a></li><li><a href="#Distributions.logpdf-Tuple{Findr.LBeta, Real}"><code>Distributions.logpdf</code></a></li><li><a href="#Distributions.pdf-Tuple{Findr.LBeta, Real}"><code>Distributions.pdf</code></a></li><li><a href="#Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat"><code>Findr.findr</code></a></li><li><a href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a></li><li><a href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a></li><li><a href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a></li><li><a href="../inference/#Findr.findr"><code>Findr.findr</code></a></li><li><a href="#Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a></li><li><a href="#Findr.fit_kde-Tuple{Any}"><code>Findr.fit_kde</code></a></li><li><a href="#Findr.fit_mixdist_EM"><code>Findr.fit_mixdist_EM</code></a></li><li><a href="#Findr.fit_mixdist_KDE"><code>Findr.fit_mixdist_KDE</code></a></li><li><a href="#Findr.fit_mixdist_mom"><code>Findr.fit_mixdist_mom</code></a></li><li><a href="#Findr.fit_mom-Tuple{Any, Any, Any}"><code>Findr.fit_mom</code></a></li><li><a href="#Findr.fit_weighted-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}, StatsBase.AbstractWeights}} where T&lt;:Real"><code>Findr.fit_weighted</code></a></li><li><a href="#Findr.groupmeans-Tuple{Any, Any, Any}"><code>Findr.groupmeans</code></a></li><li><a href="#Findr.groupmeans-Tuple{Any, Any}"><code>Findr.groupmeans</code></a></li><li><a href="#Findr.llrstats_col-Tuple{Any, Any, Any}"><code>Findr.llrstats_col</code></a></li><li><a href="#Findr.llrstats_col-Tuple{Any, Any}"><code>Findr.llrstats_col</code></a></li><li><a href="#Findr.nulldist"><code>Findr.nulldist</code></a></li><li><a href="#Findr.nulllog10pval"><code>Findr.nulllog10pval</code></a></li><li><a href="#Findr.nullpdf"><code>Findr.nullpdf</code></a></li><li><a href="#Findr.nullpval"><code>Findr.nullpval</code></a></li><li><a href="#Findr.pi0est-Tuple{Any}"><code>Findr.pi0est</code></a></li><li><a href="#Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.pprob_col</code></a></li><li><a href="#Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.pprob_col</code></a></li><li><a href="#Findr.pprob_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat"><code>Findr.pprob_col</code></a></li><li><a href="#Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.realLLR_col</code></a></li><li><a href="#Findr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat"><code>Findr.realLLR_col</code></a></li><li><a href="#Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.realLLR_col</code></a></li><li><a href="#Findr.supernormalize"><code>Findr.supernormalize</code></a></li><li><a href="#StatsAPI.fit-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real"><code>StatsAPI.fit</code></a></li><li><a href="#StatsAPI.params-Tuple{Findr.LBeta}"><code>StatsAPI.params</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="Distributions.ccdf-Tuple{Findr.LBeta, Real}" href="#Distributions.ccdf-Tuple{Findr.LBeta, Real}"><code>Distributions.ccdf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ccdf(d,x)</code></pre><p>Evaluate the complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L60-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Distributions.cdf-Tuple{Findr.LBeta, Real}" href="#Distributions.cdf-Tuple{Findr.LBeta, Real}"><code>Distributions.cdf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cdf(d,x)</code></pre><p>Evaluate the cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L49-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Distributions.fit_mle-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real" href="#Distributions.fit_mle-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real"><code>Distributions.fit_mle</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit_mle(LBeta, x)</code></pre><p>Fit an <code>LBeta</code> distribution to data <code>x</code> using maximum-likelihood estimation by exploiting its relationship to the Beta distribution.</p><p>See https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L96-L102">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Distributions.logccdf-Tuple{Findr.LBeta, Real}" href="#Distributions.logccdf-Tuple{Findr.LBeta, Real}"><code>Distributions.logccdf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">logccdf(d,x)</code></pre><p>Evaluate the log complementary cumulative distribution function of an LBeta distribution using its relation to the Beta distribution. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L71-L75">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Distributions.logpdf-Tuple{Findr.LBeta, Real}" href="#Distributions.logpdf-Tuple{Findr.LBeta, Real}"><code>Distributions.logpdf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">logpdf(d,x)</code></pre><p>Evaluate the probability density function of an LBeta distribution.  </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L39-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Distributions.pdf-Tuple{Findr.LBeta, Real}" href="#Distributions.pdf-Tuple{Findr.LBeta, Real}"><code>Distributions.pdf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pdf(d,x)</code></pre><p>Evaluate the probability density function of an LBeta distribution with support on <span>$x\geq 0$</span>. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L29-L33">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame" href="#Findr.findr-Tuple{T} where T&lt;:DataFrames.AbstractDataFrame"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(dX::T) where T&lt;:AbstractDataFrame</code></pre><p>Wrapper for <code>findr(Matrix(dX))</code> when the input <code>dX</code> is in the form of a DataFrame. The output is then also wrapped in a DataFrame with the same column names. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/Findr.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat" href="#Findr.findr-Union{Tuple{Matrix{T}}, Tuple{T}} where T&lt;:AbstractFloat"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T}) where T&lt;:AbstractFloat</code></pre><p>Compute posterior probabilities for nonzero pairwise correlations between columns of input matrix <code>X</code>. The probabilities are directed (asymmetric) in the sense that they are estimated from a column-specific background distribution. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/Findr.jl#L38-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Array{T}, Array{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X1::Matrix{T},X2::Array{T},G::Array{S})  where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero causal relations from columns of input matrix <code>X2</code> to columns of input matrix <code>X1</code>. The columns of input matrix <code>G</code> are (discrete) instrumental variables for the corresponding columns in <code>X2</code>.</p><p>Posterior probabilities are computed for the following tests</p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p>which can be combined into the mediation test (<span>$P_2 P_3$</span>), the non-independence test (<span>$P_2 P_5$</span>), or Findr&#39;s legacy combination (<span>$\frac{1}{2}(P_2 P_5 + P_4)$</span>). Alternatively, individual probability matrices for all tests are returned.</p><p>All return matrices have size ncols(X1) x ncols(X2) .</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/Findr.jl#L116-L131">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T},G::Matrix{S},pairGX::Matrix{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero causal relations between columns of input matrix <code>X</code>. The probabilities are estimated for a subset of columns of <code>X</code> that have a (discrete) instrumental variable in input matrix <code>G</code>. The matching between columns of <code>X</code> and columns of <code>G</code> is given by <code>pairGX</code>, a two-column array where the first column corresponds to a column index in <code>G</code> and the second to a column index in <code>X</code>.</p><p>Posterior probabilities are computed for the following tests</p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p>which can be combined into the mediation test (<span>$P_2 P_3$</span>), the non-independence test (<span>$P_2 P_5$</span>), or Findr&#39;s legacy combination (<span>$\frac{1}{2}(P_2 P_5 + P_4)$</span>). Alternatively, individual probability matrices for all tests are returned.</p><p>All return matrices have size ncols(X) x ncols(G) .</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/Findr.jl#L85-L100">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.findr-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Matrix{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.findr</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findr(X::Matrix{T},G::Array{S}) where {T&lt;:AbstractFloat, S&lt;:Integer}</code></pre><p>Compute posterior probabilities for nonzero differential expression of colunns of input matrix <code>X</code> across groups defined by one or more categorical variables (columns of <code>G</code>).</p><p>Return a matrix of size ncols(X) x ncols(G)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/Findr.jl#L66-L72">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_kde-Tuple{Any}" href="#Findr.fit_kde-Tuple{Any}"><code>Findr.fit_kde</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit_kde(llr)</code></pre><p>Fit a distribution function to a vector of log-likelihood ratio values <code>llr</code> using kernel density estimation. To avoid boundary effects in the KDE, the log-likelihoods (which take values in <span>$[0,\infty)$</span>) are first transformed to a vector of <code>z</code>,</p><p><span>$z = \log \left( e^{2 LLR} - 1 \right)$</span></p><p>which takes values in <span>$(-\infty,\infty)$</span>. KDE is applied to <code>z</code>, and the probability density function (pdf) for <code>llr</code> is obtained from the pdf for <code>z</code> by the usual transformation rule for functions of random variables.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L237-L247">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_mixdist_EM" href="#Findr.fit_mixdist_EM"><code>Findr.fit_mixdist_EM</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mixdist_EM(llr,ns,ng=1,test=:corr; maxiter=1000, tol=1e-3)</code></pre><p>Fit a two-component mixture distribution of two LBeta distributions to a vector of log-likelihood ratios <code>llr</code> using an EM algorithm. The first component is the true null distribution for a given Findr <code>test</code> with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The second component is the alternative distribution, assumed to follow an LBeta distribution. The prior probability <code>pi0</code> of an observation belonging to the null component is fixed and determined by the <code>pi0est</code> function. Hence only the parameters of the alternative component need to be estimated.</p><p>The EM algorithm outputs posterior probabilities of the alternative hypothesis being true, in the form of the estimated recognition distribution. The optional parameters <code>maxiter</code> (default value 1000) and <code>tol</code> (default value 1e-3) control the convergence of the EM algorithm.</p><p>The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L135-L151">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_mixdist_KDE" href="#Findr.fit_mixdist_KDE"><code>Findr.fit_mixdist_KDE</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mixdist_KDE(llr,ns,[ng,test])</code></pre><p>Return posterior probabilities for a vector of log-likelihood ratio values <code>llr</code> for a given Findr <code>test</code> with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L199-L211">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_mixdist_mom" href="#Findr.fit_mixdist_mom"><code>Findr.fit_mixdist_mom</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_mixdist_mom(llr,ns,ng=1,test=:corr)</code></pre><p>Fit a two-component mixture distribution of two LBeta distributions to a vector of log-likelihood ratios <code>llr</code> using a method-of-moments algorithm. The first component is the true null distribution for a given Findr <code>test</code> with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The second component is the alternative distribution, assumed to follow an LBeta distribution. The prior probability <code>pi0</code> of an observation belonging to the null component is fixed and determined by the <code>pi0est</code> function. Hence only the parameters of the alternative component need to be estimated.</p><p>The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L73-L87">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_mom-Tuple{Any, Any, Any}" href="#Findr.fit_mom-Tuple{Any, Any, Any}"><code>Findr.fit_mom</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit_mom(LBeta, m1, m2)</code></pre><p>Fit an <code>LBeta</code> distribution to given first and second moments <code>m1</code> and <code>m2</code> of the corresponding Beta distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L110-L114">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.fit_weighted-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}, StatsBase.AbstractWeights}} where T&lt;:Real" href="#Findr.fit_weighted-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}, StatsBase.AbstractWeights}} where T&lt;:Real"><code>Findr.fit_weighted</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit_weighted(LBeta, x, w)</code></pre><p>Fit an <code>LBeta</code> distribution to data <code>x</code> where each observation in <code>x</code> has a weight <code>w</code> between 0 and 1, using the method of moments. The function exploits the relationship to the Beta distribution and is a simple adaptation of the <code>fit</code> function in</p><p>https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L121-L127">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.groupmeans-Tuple{Any, Any, Any}" href="#Findr.groupmeans-Tuple{Any, Any, Any}"><code>Findr.groupmeans</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">groupmeans(Y,Ycol,E)</code></pre><p>Compute the size and mean of each column of matrix <code>Y</code> and of the column vector <code>Ycol</code> for each of the groups (unique values) in categorical vector <code>E</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L142-L146">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.groupmeans-Tuple{Any, Any}" href="#Findr.groupmeans-Tuple{Any, Any}"><code>Findr.groupmeans</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">groupmeans(Y,E)</code></pre><p>Compute the size and mean of each column of matrix <code>Y</code> for each of the groups (unique values) in categorical vector <code>E</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L161-L165">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.llrstats_col-Tuple{Any, Any, Any}" href="#Findr.llrstats_col-Tuple{Any, Any, Any}"><code>Findr.llrstats_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">llrstats_col(Y,Ycol,E)</code></pre><p>Compute the sufficient statistics to compute the log-likelihood ratios for Findr tests 2-5  for a given column vector <code>Ycol</code> (gene) of gene expression values with categorical instrument <code>E</code> against all columns of <code>Y</code>.</p><p><code>Ycol</code> and <code>Y</code> are assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>The sufficient statistics are:</p><ul><li>the covariance <code>ρ</code> between the given <code>Ycol</code> and all columns of <code>Y</code></li><li>the weighted average variances <code>σ1</code> of each column of matrix <code>Y</code> over the groups (unique values) in <code>E</code></li><li>the weighted average covariance <code>σ2</code> between <code>Ycol</code> and all  columns of <code>Y</code> over the groups (unique values) of <code>E</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L96-L108">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.llrstats_col-Tuple{Any, Any}" href="#Findr.llrstats_col-Tuple{Any, Any}"><code>Findr.llrstats_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">llrstats_col(Y,E)</code></pre><p>Compute the sufficient statistics to compute the log-likelihood ratios for Findr tests 2  for a given categorical vector <code>E</code> against all columns of <code>Y</code>.</p><p><code>Y</code> is assumed to have undergone supernormalization with each col having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p><p>The sufficient statistics are the weighted average variances <code>σ1</code> of each column of matrix <code>Y</code> over the groups (unique values) in <code>E</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L124-L132">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.nulldist" href="#Findr.nulldist"><code>Findr.nulldist</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nulldist(ns,[ng,test])</code></pre><p>Return an LBeta distribution that is the null distribution of the log-likelihood ratio for a given Findr test with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With only one input argument, the null distribution for the correlation test with <code>ns</code> samples is returned. With two input arguments, or  with three arguments and <code>test</code> equal to &quot;corr&quot;, the null distribution for the correlation test with <code>ns</code> samples is returned and the second argument is ignored</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/randomLLR.jl#L2-L14">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.nulllog10pval" href="#Findr.nulllog10pval"><code>Findr.nulllog10pval</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nulllog10pval(llr,ns,[ng,test])</code></pre><p>Return negative log10 p-values for a vector of log-likelihood ratio values <code>llr</code> under the null distribution of the log-likelihood ratio for a given Findr test with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/randomLLR.jl#L53-L65">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.nullpdf" href="#Findr.nullpdf"><code>Findr.nullpdf</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nullpdf(llr,ns,[ng,test])</code></pre><p>Return probability distribution function evaluations for a vector of log-likelihood ratio values <code>llr</code> under the null distribution of the log-likelihood ratio for a given Findr test with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/randomLLR.jl#L73-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.nullpval" href="#Findr.nullpval"><code>Findr.nullpval</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nullpval(llr,ns,[ng,test])</code></pre><p>Return p-values for a vector of log-likelihood ratio values <code>llr</code> under the null distribution of the log-likelihood ratio for a given Findr test with sample size <code>ns</code> and number of genotype groups <code>ng</code>. The input variable <code>test</code> can take the values:</p><ul><li>&#39;:corr&#39; - <strong>correlation test</strong> (test 0)</li><li>&#39;:link&#39; - <strong>linkage test</strong> (test 1/2)</li><li>&#39;:med&#39; - <strong>mediation test</strong> (test 3)</li><li>&#39;:relev&#39; - <strong>relevance test</strong> (test 4)</li><li>&#39;:pleio&#39; - <strong>pleiotropy test</strong> (test 5)</li></ul><p>With two input arguments, the correlation test with <code>ns</code> samples is used. With three input arguments, or with four arguments and <code>test</code> equal to &quot;:corr&quot;, the correlation test with <code>ns</code> samples is used and the third argument is ignored.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/randomLLR.jl#L31-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.pi0est-Tuple{Any}" href="#Findr.pi0est-Tuple{Any}"><code>Findr.pi0est</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pi0est(pval)</code></pre><p>Estimate the proportion π0 of truly null features in a vector <code>pval</code> of p-values using Storey&#39;s method</p><p>See also http://varianceexplained.org/files/pi0boot.pdf</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L260-L266">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.pprob_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pprob_col(Y,E)</code></pre><p>Compute the posterior probabilities for differential expression of columns of matrix <code>Y</code> in the groups defined by  categorical vector <code>E</code> using Findr test 2 (<strong>Linkage test</strong>) </p><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L52-L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.pprob_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.pprob_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pprob_col(Y,Ycol,E)</code></pre><p>Compute the posterior probabilities for the Findr causal tests for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of matrix <code>Y</code>: </p><pre><code class="nohighlight hljs">- Test 2 (**Linkage test**) 
- Test 3 (**Mediation test**)
- Test 4 (**Relevance test**)
- Test 5 (**Pleiotropy test**)</code></pre><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one.</p><p>For test 2, 4, and 5 the posterior probabilities are the probabilities of the alternative hypothesis being true. For test 3 they are the probabilities of the null hypothesis being true.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L18-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.pprob_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat" href="#Findr.pprob_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat"><code>Findr.pprob_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pprob_col(Y,Ycol)</code></pre><p>Compute the posterior probabilities for Findr test 0 (<strong>correlation test</strong>) for a given column vector <code>Ycol</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/posteriorprobs.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y,Ycol,E)</code></pre><p>Compute the log-likelihood ratios for Findr test 2 (<strong>Linkage test</strong>)  for a given categorical vector <code>E</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> is assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L51-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}" href="#Findr.realLLR_col-Union{Tuple{S}, Tuple{T}, Tuple{Matrix{T}, Vector{T}, Vector{S}}} where {T&lt;:AbstractFloat, S&lt;:Integer}"><code>Findr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y,Ycol,E)</code></pre><p>Compute the log-likelihood ratios for the Findr causal tests for a given column vector <code>Ycol</code> with categorical instrument <code>E</code> against all columns of matrix <code>Y</code> : </p><ul><li>Test 2 (<strong>Linkage test</strong>) </li><li>Test 3 (<strong>Mediation test</strong>)</li><li>Test 4 (<strong>Relevance test</strong>)</li><li>Test 5 (<strong>Pleiotropy test</strong>)</li></ul><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L14-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat" href="#Findr.realLLR_col-Union{Tuple{T}, Tuple{Matrix{T}, Vector{T}}} where T&lt;:AbstractFloat"><code>Findr.realLLR_col</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">realLLR_col(Y,Ycol)</code></pre><p>Compute the log-likelihood ratios for Findr test 0 (<strong>correlation test</strong>) for a given column vector <code>Ycol</code> against all columns of matrix <code>Y</code>.</p><p><code>Y</code> and <code>Ycol</code> are assumed to have undergone supernormalization with each column having mean zero and variance one. The LLRs are scaled by the number of rows (samples).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/realLLR.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Findr.supernormalize" href="#Findr.supernormalize"><code>Findr.supernormalize</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">supernormalize(X[, c])</code></pre><p>Convert each column of matrix or dataframe <code>X</code> of reals into standard normally distributed values using a rank-based inverse normal transformation. Then scale each column to have variance one.</p><p>Note that after the inverse normal transformation, each column has mean zero and identical variance (if we use ordinal ranking). Hence rescaling can be done once on the whole matrix.</p><p>The formula and default value for the paramater <code>c</code> come from this paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2921808/</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/supernormalization.jl#L1-L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="StatsAPI.fit-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real" href="#StatsAPI.fit-Union{Tuple{T}, Tuple{Type{&lt;:Findr.LBeta}, AbstractArray{T}}} where T&lt;:Real"><code>StatsAPI.fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(LBeta, x)</code></pre><p>Fit an <code>LBeta</code> distribution to data <code>x</code> using the method of moments by exploiting its relationship to the Beta distribution.</p><p>See https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L82-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="StatsAPI.params-Tuple{Findr.LBeta}" href="#StatsAPI.params-Tuple{Findr.LBeta}"><code>StatsAPI.params</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">params(d)</code></pre><p>Get the parameters of an LBeta distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/tmichoel/Findr.jl/blob/2cc99b499a74f33308525fd74fe50c9d351aa66a/src/lbeta.jl#L21-L25">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../testLLR/">« Tests to evaluate</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 30 May 2023 20:06">Tuesday 30 May 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
